{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "view-in-github"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/wikipathways/BioThings_Explorer_PFOCR_clustering/blob/main/bte_clustering_AA.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Environment setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#install the required packages\n",
    "!pip install numpy pandas requests requests_cache SetSimilaritySearch fisher"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "BrREAxe69lsV"
   },
   "outputs": [],
   "source": [
    "#load packages\n",
    "from urllib.request import urlopen\n",
    "import requests\n",
    "import json\n",
    "from copy import copy,deepcopy\n",
    "import pandas as pd\n",
    "from fisher import pvalue_npy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## User Inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# which node ID from message.query_graph.nodes should be used for grouping. \n",
    "# The entities mapped to that node ID are the \"result entities\"\n",
    "# If all then use all the node ids in query\n",
    "# If specific node id set (e.g. {'MESH:D000068877', 'MESH:D001249'}) then use only these node ids\n",
    "node_ID_for_grouping = \"all\"\n",
    "\n",
    "# n corresponds to the number of results desired\n",
    "n = 15\n",
    "\n",
    "# user-specified node IDs for the subject, the object, both, or neither that must be included in the figures\n",
    "# If all then use all the node ids in query\n",
    "# If specific node id set (e.g. {'MESH:D000068877', 'MESH:D001249'}) then use only these node ids\n",
    "required_curies = \"all\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get BTE TRAPI Results\n",
    "\n",
    "Master notebook: https://github.com/wikipathways/pathway-figure-ocr/blob/master/notebooks/bte_clustering.ipynb\n",
    "\n",
    "This notebook reads in the TRAPI results from a URL for the query below.\n",
    "\n",
    "### Query\n",
    "Imatinib - [Gene] - Asthma\n",
    "\n",
    "### Results URL\n",
    "https://arax.ncats.io/api/arax/v1.3/response/7b14f961-9066-41f7-9e3b-d76b2b4a7fac (83kB, 7 results); \"results\" in n1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'logs': [{'code': None, 'level': 'INFO', 'message': 'job status available at: http://bte.transltr.io/v1/check_query_status/9uTnf0I7u2', 'timestamp': '2022-12-15T21:32:44.148Z'}, {'code': None, 'level': 'WARNING', 'message': 'The following APIs were unavailable at the time of execution: Service Provider TRAPI, SRI reference KG (ITRB DEV) (trapi v-1.3.0), SRI reference KG (ITRB DEV) (trapi v-1.3.0), SRI Reference Knowledge Graph API (trapi v-1.3.0), CAM-KP API, Automat-mychem-info(Trapi v1.3.0), Knowledge Collaboratory API, CAM-KP API, CAM-KP API', 'timestamp': '2022-12-15T21:32:39.598Z'}, {'code': None, 'level': 'INFO', 'message': 'Node n0 with id [MESH:D000068877] and category [biolink:Drug] augmented with category [biolink:MolecularMixture] inferred from id.', 'timestamp': '2022-12-15T21:32:39.681Z'}, {'code': None, 'level': 'INFO', 'message': 'Node n2 with id [MESH:D001249] and category [biolink:Disease] augmented with category [biolink:PhenotypicFeature] inferred from id.', 'timestamp': '2022-12-15T21:32:39.703Z'}, {'code': None, 'level': 'DEBUG', 'message': 'BTE identified 3 qNodes from your query graph', 'timestamp': '2022-12-15T21:32:39.703Z'}, {'code': None, 'level': 'DEBUG', 'message': 'BTE identified 2 qEdges from your query graph', 'timestamp': '2022-12-15T21:32:39.707Z'}, {'code': None, 'level': 'INFO', 'message': 'Executing e1 (reversed): n1 <-- n2', 'timestamp': '2022-12-15T21:32:39.923Z'}, {'code': None, 'level': 'DEBUG', 'message': 'REDIS cache is  enabled.', 'timestamp': '2022-12-15T21:32:39.946Z'}, {'code': None, 'level': 'DEBUG', 'message': 'BTE finds cached records for e1', 'timestamp': '2022-12-15T21:32:40.585Z'}, {'code': None, 'level': 'INFO', 'message': 'e1 execution: 0 queries (0 success/0 fail) and (1) cached qEdges return (4151) records', 'timestamp': '2022-12-15T21:32:40.591Z'}, {'code': None, 'level': 'INFO', 'message': 'Executing e0: n0 --> n1', 'timestamp': '2022-12-15T21:32:43.296Z'}, {'code': None, 'level': 'DEBUG', 'message': 'REDIS cache is  enabled.', 'timestamp': '2022-12-15T21:32:43.317Z'}, {'code': None, 'level': 'DEBUG', 'message': 'BTE finds cached records for e0', 'timestamp': '2022-12-15T21:32:43.326Z'}, {'code': None, 'level': 'INFO', 'message': 'e0 execution: 0 queries (0 success/0 fail) and (1) cached qEdges return (37) records', 'timestamp': '2022-12-15T21:32:43.326Z'}, {'code': None, 'level': 'DEBUG', 'message': 'Edge manager is managing 2 qEdges.', 'timestamp': '2022-12-15T21:32:39.923Z'}, {'code': None, 'level': 'DEBUG', 'message': \"Edge manager is sending next qEdge 'e1' for execution.\", 'timestamp': '2022-12-15T21:32:39.923Z'}, {'code': None, 'level': 'DEBUG', 'message': \"'e1' kept (4151) / dropped (0) records.\", 'timestamp': '2022-12-15T21:32:41.328Z'}, {'code': None, 'level': 'DEBUG', 'message': 'Next qEdge will pick lower entity value to use for query.', 'timestamp': '2022-12-15T21:32:43.296Z'}, {'code': None, 'level': 'DEBUG', 'message': \"Edge manager is sending next qEdge 'e0' for execution.\", 'timestamp': '2022-12-15T21:32:43.296Z'}, {'code': None, 'level': 'DEBUG', 'message': \"'e0' kept (11) / dropped (26) records.\", 'timestamp': '2022-12-15T21:32:43.384Z'}, {'code': None, 'level': 'DEBUG', 'message': \"'e1' kept (22) / dropped (4129) records.\", 'timestamp': '2022-12-15T21:32:43.547Z'}, {'code': None, 'level': 'DEBUG', 'message': \"'e0' kept (11) / dropped (0) records.\", 'timestamp': '2022-12-15T21:32:43.549Z'}, {'code': None, 'level': 'INFO', 'message': \"'e0' keeps (11) records!\", 'timestamp': '2022-12-15T21:32:43.551Z'}, {'code': None, 'level': 'INFO', 'message': \"'e1' keeps (22) records!\", 'timestamp': '2022-12-15T21:32:43.552Z'}, {'code': None, 'level': 'DEBUG', 'message': 'Edge manager collected (33) records!', 'timestamp': '2022-12-15T21:32:43.552Z'}, {'code': None, 'level': 'DEBUG', 'message': \"Successfully scored 4 results, couldn't score 3 results.\", 'timestamp': '2022-12-15T21:32:44.147Z'}, {'code': None, 'level': 'INFO', 'message': 'Execution Summary: (9) nodes / (24) edges / (7) results; (0/0) queries (2 cached qEdges) returned results from (0) unique APIs ', 'timestamp': '2022-12-15T21:32:44.148Z'}, {'code': None, 'level': 'INFO', 'message': 'APIs: ', 'timestamp': '2022-12-15T21:32:44.148Z'}], 'message': {'knowledge_graph': {'edges': {'02217ffa55351b0c3c0a32431158ca27': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:increases_secretion_of', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:3576', 'predicate': 'biolink:increases_secretion_of', 'subject': 'PUBCHEM.COMPOUND:123596'}, '05ec582ef37cc52970d0549edb0da775': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:increases_response_to', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:5156', 'predicate': 'biolink:response_affected_by', 'subject': 'PUBCHEM.COMPOUND:123596'}, '133406fa1089f0630ad505ba930f0615': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:disgenet'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:mydisease-info'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:publications', 'value': ['PMID:24152156']}, {'attribute_type_id': 'disgenet-DPI', 'value': 0.923}, {'attribute_type_id': 'disgenet-DSI', 'value': 0.266}, {'attribute_type_id': 'disgenet-EI', 'value': 1}, {'attribute_type_id': 'disgenet-score', 'value': 0.01}, {'attribute_type_id': 'disgenet_source', 'value': 'BEFREE'}], 'object': 'HP:0002099', 'predicate': 'biolink:gene_associated_with_condition', 'subject': 'NCBIGene:7422'}, '17a4c659b4a9f1a8bcd164cd17c03db6': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:pfocr'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-pfocr'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'figure_download_url', 'value': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4973429/bin/fphar-07-00237-g0002.jpg'}, {'attribute_type_id': 'figure_title', 'value': 'Involvement of AKR1B1 in pathophysiology'}, {'attribute_type_id': 'pmc_reference', 'value': 'PMC4973429'}], 'object': 'HP:0002099', 'predicate': 'biolink:occurs_together_in_literature_with', 'subject': 'NCBIGene:3725'}, '1a199ceb3b74961d7b72e222ac9b3172': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:semmeddb'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-semmeddb'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:publications', 'value': ['PMID:28024734']}, {'attribute_type_id': 'biolink:original_object', 'value': 'C0004096'}, {'attribute_type_id': 'biolink:original_predicate', 'value': 'ASSOCIATED_WITH'}, {'attribute_type_id': 'biolink:original_subject', 'value': 'C0812307'}, {'attribute_type_id': 'original_object_name', 'value': 'Asthma'}, {'attribute_type_id': 'original_subject_name', 'value': 'LYN gene'}], 'object': 'HP:0002099', 'predicate': 'biolink:associated_with', 'subject': 'NCBIGene:4067'}, '23a8bbe1078a780814d37f2785fad1ab': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:increases_secretion_of', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:7422', 'predicate': 'biolink:increases_secretion_of', 'subject': 'PUBCHEM.COMPOUND:123596'}, '3b84553e4e9b9466e82c594cc5913b49': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:decreases_response_to', 'value_type_id': 'EDAM:data_0006', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}], 'object': 'NCBIGene:4067', 'predicate': 'biolink:response_decreased_by', 'subject': 'PUBCHEM.COMPOUND:123596'}, '3bdac66d7282f313bc444533ecb84a4f': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:original_knowledge_source', 'value': 'infores:text-mining-provider-targeted', 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_data_source', 'value': 'infores:pubmed', 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:has_evidence_count', 'value': 3, 'value_type_id': 'biolink:EvidenceCount'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:tmkp_confidence_score', 'value': 0.99767165, 'value_type_id': 'biolink:ConfidenceLevel'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document', 'value': 'PMID:19196819|PMID:19196819|PMID:19196819', 'value_type_id': 'biolink:Publication'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_study_result', 'attributes': [{'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_text', 'value': 'VEGF polymorphisms are associated with childhood asthma, lung function and airway responsiveness in two populations, suggesting that VEGF polymorphisms influence asthma susceptibility, airflow obstruction and airways responsiveness.', 'value_type_id': 'EDAM:data_3671'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document', 'value': 'PMID:19196819', 'value_type_id': 'biolink:Publication', 'value_url': 'https://pubmed.ncbi.nlm.nih.gov/19196819/'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_text_located_in', 'value': 'abstract', 'value_type_id': 'IAO_0000314'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:extraction_confidence_score', 'value': 0.9979881, 'value_type_id': 'EDAM:data_1772'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:subject_location_in_text', 'value': '0|4', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:text-mining-provider-targeted ', 'attribute_type_id': 'biolink:object_location_in_text', 'value': '49|55', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document_year', 'value': 2009, 'value_type_id': 'UO:0000036'}], 'value': 'tmkp:0e7972a590450fdf5d10adaa72434ef7fa501362e9d172499c1735d77e13ad2d', 'value_type_id': 'biolink:TextMiningResult'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_study_result', 'attributes': [{'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_text', 'value': 'VEGF polymorphisms are associated with childhood asthma, lung function and airway responsiveness in two populations, suggesting that VEGF polymorphisms influence asthma susceptibility, airflow obstruction and airways responsiveness.', 'value_type_id': 'EDAM:data_3671'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document', 'value': 'PMID:19196819', 'value_type_id': 'biolink:Publication', 'value_url': 'https://pubmed.ncbi.nlm.nih.gov/19196819/'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_text_located_in', 'value': 'abstract', 'value_type_id': 'IAO_0000314'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:extraction_confidence_score', 'value': 0.9977518, 'value_type_id': 'EDAM:data_1772'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:subject_location_in_text', 'value': '133|137', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:text-mining-provider-targeted ', 'attribute_type_id': 'biolink:object_location_in_text', 'value': '49|55', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document_year', 'value': 2009, 'value_type_id': 'UO:0000036'}], 'value': 'tmkp:338c6421377e7f6ef40197260064a6d7015bde00a8cb774fd97c3c385b9e9904', 'value_type_id': 'biolink:TextMiningResult'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_study_result', 'attributes': [{'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_text', 'value': 'Association of VEGF polymorphisms with childhood asthma, lung function and airway responsiveness.', 'value_type_id': 'EDAM:data_3671'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document', 'value': 'PMID:19196819', 'value_type_id': 'biolink:Publication', 'value_url': 'https://pubmed.ncbi.nlm.nih.gov/19196819/'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_text_located_in', 'value': 'title', 'value_type_id': 'IAO_0000314'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:extraction_confidence_score', 'value': 0.99727505, 'value_type_id': 'EDAM:data_1772'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:subject_location_in_text', 'value': '15|19', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:text-mining-provider-targeted ', 'attribute_type_id': 'biolink:object_location_in_text', 'value': '49|55', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document_year', 'value': 2009, 'value_type_id': 'UO:0000036'}], 'value': 'tmkp:10fd1e6a9b3c224a0813d96e4e928c65327c82b2385164c4cf7c82456c21b097', 'value_type_id': 'biolink:TextMiningResult'}], 'object': 'HP:0002099', 'predicate': 'biolink:contributes_to', 'subject': 'NCBIGene:7422'}, '47b528f4619fe13a988467e04b78b536': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:increases_activity_of', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:3725', 'predicate': 'biolink:increases_activity_of', 'subject': 'PUBCHEM.COMPOUND:123596'}, '651e711baac3604af1b09d0e92751d2d': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:disgenet'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:mydisease-info'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:publications', 'value': ['PMID:18823309']}, {'attribute_type_id': 'disgenet-DPI', 'value': 0.923}, {'attribute_type_id': 'disgenet-DSI', 'value': 0.404}, {'attribute_type_id': 'disgenet-EI', 'value': 1}, {'attribute_type_id': 'disgenet-score', 'value': 0.01}, {'attribute_type_id': 'disgenet_source', 'value': 'BEFREE'}], 'object': 'HP:0002099', 'predicate': 'biolink:gene_associated_with_condition', 'subject': 'NCBIGene:841'}, '6ba3d9ff480ec75e66a38889f0e9c4b2': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:increases_degradation_of', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:142', 'predicate': 'biolink:increases_degradation_of', 'subject': 'PUBCHEM.COMPOUND:123596'}, '79c5be38756c7936c6706468d30a3133': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:decreases_secretion_of', 'value_type_id': 'EDAM:data_0006', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}], 'object': 'NCBIGene:7422', 'predicate': 'biolink:decreases_secretion_of', 'subject': 'PUBCHEM.COMPOUND:123596'}, '7ee912dbb014b53ed3fcea0e7dc883f8': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:monarchinitiative'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biolink-api'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:publications', 'value': ['PMID:12353035']}, {'attribute_type_id': 'biolink:original_predicate', 'value': 'has phenotype'}, {'attribute_type_id': 'monarch_source_database', 'value': ['https://archive.monarchinitiative.org/#orphanet', 'https://archive.monarchinitiative.org/#hpoa']}], 'object': 'HP:0002099', 'predicate': 'biolink:gene_associated_with_condition', 'subject': 'NCBIGene:841'}, '81900675077f00cbc42b8d26d95589fc': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:semmeddb'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-semmeddb'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:publications', 'value': ['PMID:18823309']}, {'attribute_type_id': 'biolink:original_object', 'value': 'C0004096'}, {'attribute_type_id': 'biolink:original_predicate', 'value': 'ASSOCIATED_WITH'}, {'attribute_type_id': 'biolink:original_subject', 'value': 'C1332667'}, {'attribute_type_id': 'original_object_name', 'value': 'Asthma'}, {'attribute_type_id': 'original_subject_name', 'value': 'CASP8 gene'}], 'object': 'HP:0002099', 'predicate': 'biolink:associated_with', 'subject': 'NCBIGene:841'}, '83f80061137b13dfde6941f711ab53eb': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:decreases_expression_of', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:5156', 'predicate': 'biolink:decreases_expression_of', 'subject': 'PUBCHEM.COMPOUND:123596'}, '8b7bd93b93c73afa7122d4b6f11c9258': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:pfocr'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-pfocr'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'figure_download_url', 'value': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8065335/bin/41392_2021_568_Fig5_HTML.jpg'}, {'attribute_type_id': 'figure_title', 'value': 'Putative mechanisms for CVL218 as a PARP1 inhibitor to combat the COVID-19 disease, derived based on the data present in this study and the known antiviral activities of PARP1 inhibitors previously reported in the literature'}, {'attribute_type_id': 'pmc_reference', 'value': 'PMC8065335'}], 'object': 'HP:0002099', 'predicate': 'biolink:occurs_together_in_literature_with', 'subject': 'NCBIGene:142'}, 'a823bd04f08a5591fba87dc56d52647d': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:increases_activity_of', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:841', 'predicate': 'biolink:increases_activity_of', 'subject': 'PUBCHEM.COMPOUND:123596'}, 'ac85a6ff34dc2170e13fe0019d983c85': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:pfocr'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-pfocr'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'figure_download_url', 'value': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8065335/bin/41392_2021_568_Fig5_HTML.jpg'}, {'attribute_type_id': 'figure_title', 'value': 'Putative mechanisms for CVL218 as a PARP1 inhibitor to combat the COVID-19 disease, derived based on the data present in this study and the known antiviral activities of PARP1 inhibitors previously reported in the literature'}, {'attribute_type_id': 'pmc_reference', 'value': 'PMC8065335'}], 'object': 'HP:0002099', 'predicate': 'biolink:occurs_together_in_literature_with', 'subject': 'NCBIGene:3576'}, 'bca736d8e683b35a133bee44f2a75629': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:increases_response_to', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:5156', 'predicate': 'biolink:response_increased_by', 'subject': 'PUBCHEM.COMPOUND:123596'}, 'c2323af968047313b3f7d025c3f7c804': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:decreases_response_to', 'value_type_id': 'EDAM:data_0006', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}], 'object': 'NCBIGene:4067', 'predicate': 'biolink:response_affected_by', 'subject': 'PUBCHEM.COMPOUND:123596'}, 'c80d1f9d65f26b19dec2ca1395d1988e': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:semmeddb'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-semmeddb'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:publications', 'value': ['PMID:18421428']}, {'attribute_type_id': 'biolink:original_object', 'value': 'C0004096'}, {'attribute_type_id': 'biolink:original_predicate', 'value': 'ASSOCIATED_WITH'}, {'attribute_type_id': 'biolink:original_subject', 'value': 'C1171892'}, {'attribute_type_id': 'original_object_name', 'value': 'Asthma'}, {'attribute_type_id': 'original_subject_name', 'value': 'VEGF protein, human'}], 'object': 'HP:0002099', 'predicate': 'biolink:associated_with', 'subject': 'NCBIGene:7422'}, 'e0cd55dd7bf8840506b2dab758bfcd4e': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:original_knowledge_source', 'value': 'infores:text-mining-provider-targeted', 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_data_source', 'value': 'infores:pubmed', 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:has_evidence_count', 'value': 1, 'value_type_id': 'biolink:EvidenceCount'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:tmkp_confidence_score', 'value': 0.9979808, 'value_type_id': 'biolink:ConfidenceLevel'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document', 'value': 'PMID:25530067', 'value_type_id': 'biolink:Publication'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_study_result', 'attributes': [{'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_text', 'value': 'Compared with patients with asthma attacks, patients with exacerbated COPD had increased IL-8 expression in both serum and EBC.', 'value_type_id': 'EDAM:data_3671'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document', 'value': 'PMID:25530067', 'value_type_id': 'biolink:Publication', 'value_url': 'https://pubmed.ncbi.nlm.nih.gov/25530067/'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_text_located_in', 'value': 'abstract', 'value_type_id': 'IAO_0000314'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:extraction_confidence_score', 'value': 0.9979808, 'value_type_id': 'EDAM:data_1772'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:subject_location_in_text', 'value': '89|93', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:text-mining-provider-targeted ', 'attribute_type_id': 'biolink:object_location_in_text', 'value': '28|34', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document_year', 'value': 2014, 'value_type_id': 'UO:0000036'}], 'value': 'tmkp:2e332a4f0a754fd9087e3730db021775f5b4815d37a79ab2b3d5532fc6f5b828', 'value_type_id': 'biolink:TextMiningResult'}], 'object': 'HP:0002099', 'predicate': 'biolink:contributes_to', 'subject': 'NCBIGene:3576'}, 'e2e73e88ecdce75377ff36aae8c52281': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:disgenet'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:mydisease-info'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:publications', 'value': ['PMID:29189668']}, {'attribute_type_id': 'disgenet-DPI', 'value': 0.962}, {'attribute_type_id': 'disgenet-DSI', 'value': 0.31}, {'attribute_type_id': 'disgenet-EI', 'value': 1}, {'attribute_type_id': 'disgenet-score', 'value': 0.01}, {'attribute_type_id': 'disgenet_source', 'value': 'BEFREE'}], 'object': 'HP:0002099', 'predicate': 'biolink:gene_associated_with_condition', 'subject': 'NCBIGene:3576'}, 'fbb3eab04ddeb56570d39d1ee2f82da3': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:monarchinitiative'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biolink-api'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:original_predicate', 'value': 'has phenotype'}, {'attribute_type_id': 'monarch_source_database', 'value': ['https://archive.monarchinitiative.org/#omim', 'https://archive.monarchinitiative.org/#hpoa']}], 'object': 'HP:0002099', 'predicate': 'biolink:gene_associated_with_condition', 'subject': 'NCBIGene:5156'}}, 'nodes': {'HP:0002099': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['HP:0002099', 'NCIT:C113673', 'UMLS:C0004096', 'UMLS:C0085129', 'UMLS:C0340062', 'UMLS:C3714497', 'MEDDRA:10003553', 'MEDDRA:10003555', 'MEDDRA:10003560', 'MEDDRA:10003561', 'MEDDRA:10003565', 'MEDDRA:10006450', 'MEDDRA:10037993', 'MEDDRA:10056285', 'MEDDRA:10066091', 'MEDDRA:10066863', 'SNOMEDCT:195967001', 'SNOMEDCT:991000119106', 'MESH:D001249', 'MESH:D016535']}, {'attribute_type_id': 'biolink:synonym', 'value': ['Asthma', 'Reactive Airway Disease', 'Bronchial Hyperreactivity', 'Hyperreactive airway disease', 'Reactive airway disease']}], 'categories': ['biolink:PhenotypicFeature'], 'name': 'Asthma'}, 'NCBIGene:142': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['NCBIGene:142', 'ENSEMBL:ENSG00000143799', 'HGNC:270', 'OMIM:173870', 'UMLS:C1538577', 'UniProtKB:A0A024R3T8', 'UniProtKB:P09874', 'PR:P09874', 'ENSEMBL:ENSP00000355759', 'UMLS:C1567710', 'UMLS:C3641724']}, {'attribute_type_id': 'biolink:synonym', 'value': ['PARP1', 'PARP1 gene', 'A0A024R3T8_HUMAN Poly [ADP-ribose] polymerase (trembl)', 'PARP1_HUMAN Poly [ADP-ribose] polymerase 1 (sprot)', 'poly [ADP-ribose] polymerase 1 (human)', 'PARP1 protein, human', 'Poly [ADP-Ribose] Polymerase 1 p85, human']}], 'categories': ['biolink:Gene'], 'name': 'PARP1'}, 'NCBIGene:3576': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['NCBIGene:3576', 'ENSEMBL:ENSG00000169429', 'HGNC:6025', 'OMIM:146930', 'UMLS:C1366571', 'UniProtKB:A0A024RDA5', 'UniProtKB:P10145', 'PR:P10145', 'ENSEMBL:ENSP00000306512', 'UMLS:C1698756']}, {'attribute_type_id': 'biolink:synonym', 'value': ['CXCL8', 'CXCL8 gene', 'A0A024RDA5_HUMAN Multifunctional fusion protein (trembl)', 'IL8_HUMAN Interleukin-8 (sprot)', 'interleukin-8 (human)', 'CXCL8 protein, human']}], 'categories': ['biolink:Gene'], 'name': 'CXCL8'}, 'NCBIGene:3725': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['NCBIGene:3725', 'ENSEMBL:ENSG00000177606', 'HGNC:6204', 'OMIM:165160', 'UMLS:C0162508', 'UniProtKB:P05412', 'PR:P05412', 'ENSEMBL:ENSP00000360266', 'ENSEMBL:ENSP00000503132', 'UMLS:C3536909']}, {'attribute_type_id': 'biolink:synonym', 'value': ['JUN', 'JUN gene', 'JUN_HUMAN Transcription factor AP-1 (sprot)', 'transcription factor AP-1 (human)', 'Transcription Factor AP-1, human']}], 'categories': ['biolink:Gene'], 'name': 'JUN'}, 'NCBIGene:4067': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['NCBIGene:4067', 'ENSEMBL:ENSG00000254087', 'HGNC:6735', 'OMIM:165120', 'UMLS:C0812307', 'UniProtKB:A8K379', 'UniProtKB:P07948', 'PR:P07948', 'UMLS:C1442804', 'UniProtKB:Q6NUK7']}, {'attribute_type_id': 'biolink:synonym', 'value': ['LYN', 'LYN gene', 'A8K379_HUMAN Tyrosine-protein kinase (trembl)', 'LYN_HUMAN Tyrosine-protein kinase Lyn (sprot)', 'tyrosine-protein kinase Lyn (human)', 'LYN protein, human', 'Q6NUK7_HUMAN Tyrosine-protein kinase (Fragment) (trembl)']}], 'categories': ['biolink:Gene'], 'name': 'LYN'}, 'NCBIGene:5156': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['NCBIGene:5156', 'ENSEMBL:ENSG00000134853', 'HGNC:8803', 'OMIM:173490', 'UMLS:C1335201', 'UniProtKB:P16234', 'PR:P16234', 'UMLS:C3853694']}, {'attribute_type_id': 'biolink:synonym', 'value': ['PDGFRA', 'PDGFRA gene', 'PGFRA_HUMAN Platelet-derived growth factor receptor alpha (sprot)', 'platelet-derived growth factor receptor alpha (human)', 'Platelet-Derived Growth Factor Receptor Alpha, human']}], 'categories': ['biolink:Gene'], 'name': 'PDGFRA'}, 'NCBIGene:7422': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['NCBIGene:7422', 'ENSEMBL:ENSG00000112715', 'HGNC:12680', 'OMIM:192240', 'UMLS:C1823619', 'UniProtKB:A0A024RD33', 'UniProtKB:A0A024RD37', 'UniProtKB:A0A0Y0IMM4', 'UniProtKB:A2A2V4', 'ENSEMBL:ENSP00000230480', 'UniProtKB:P15692', 'PR:P15692', 'UMLS:C0376854', 'UMLS:C0376855', 'UMLS:C0376856', 'UMLS:C0525171', 'UMLS:C0752432', 'UMLS:C1171892']}, {'attribute_type_id': 'biolink:synonym', 'value': ['VEGFA', 'VEGFA gene', 'A0A024RD33_HUMAN Vascular endothelial growth factor A (trembl)', 'A0A024RD37_HUMAN Vascular endothelial growth factor A (trembl)', 'A0A0Y0IMM4_HUMAN Vascular endothelial growth factor A (Fragment) (trembl)', 'A2A2V4_HUMAN Vascular endothelial growth factor A (trembl)', 'VEGFA_HUMAN Vascular endothelial growth factor A (sprot)', 'vascular endothelial growth factor A (human)', 'VEGF165 protein, human', 'VEGF189 protein, human', 'VEGF206 protein, human', 'VEGF145 protein, human', 'VEGF183 protein, human', 'VEGF protein, human']}], 'categories': ['biolink:Gene'], 'name': 'VEGFA'}, 'NCBIGene:841': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['NCBIGene:841', 'ENSEMBL:ENSG00000064012', 'HGNC:1509', 'OMIM:601763', 'UMLS:C1332667', 'UniProtKB:A0A024R3Z8', 'UniProtKB:Q14790', 'PR:Q14790', 'UMLS:C1144061']}, {'attribute_type_id': 'biolink:synonym', 'value': ['CASP8', 'CASP8 gene', 'A0A024R3Z8_HUMAN HCG16983, isoform CRA_b (trembl)', 'CASP8_HUMAN Caspase-8 (sprot)', 'caspase-8 (human)', 'CASP8 protein, human']}], 'categories': ['biolink:Gene'], 'name': 'CASP8'}, 'PUBCHEM.COMPOUND:123596': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['PUBCHEM.COMPOUND:123596', 'PUBCHEM.COMPOUND:90091706', 'CHEMBL.COMPOUND:CHEMBL1642', 'UNII:8A1O1M485B', 'CHEBI:31690', 'MESH:D000068877', 'CAS:220127-57-1', 'INCHIKEY:YLMAHDNUQAMNNX-UHFFFAOYSA-N', 'UMLS:C0906802', 'UMLS:C0939537']}, {'attribute_type_id': 'biolink:synonym', 'value': ['Imatinib mesylate', 'Methanesulfonate;4-[(4-methylpiperazin-4-ium-1-yl)methyl]-N-[4-methyl-3-[(4-pyridin-3-ylpyrimidin-2-yl)amino]phenyl]benzamide', 'IMATINIB MESYLATE', 'imatinib methanesulfonate', 'Imatinib Mesylate', 'STI571']}], 'categories': ['biolink:MolecularMixture'], 'name': 'Imatinib mesylate'}}}, 'query_graph': {'edges': {'e0': {'object': 'n1', 'predicates': ['biolink:related_to'], 'subject': 'n0'}, 'e1': {'object': 'n2', 'predicates': ['biolink:related_to'], 'subject': 'n1'}}, 'nodes': {'n0': {'categories': ['biolink:Drug', 'biolink:MolecularMixture'], 'ids': ['MESH:D000068877']}, 'n1': {'categories': ['biolink:Gene']}, 'n2': {'categories': ['biolink:Disease', 'biolink:PhenotypicFeature'], 'ids': ['MESH:D001249']}}}, 'results': [{'edge_bindings': {'e0': [{'id': '23a8bbe1078a780814d37f2785fad1ab'}, {'id': '79c5be38756c7936c6706468d30a3133'}], 'e1': [{'id': '133406fa1089f0630ad505ba930f0615'}, {'id': '3bdac66d7282f313bc444533ecb84a4f'}, {'id': 'c80d1f9d65f26b19dec2ca1395d1988e'}]}, 'node_bindings': {'n0': [{'id': 'PUBCHEM.COMPOUND:123596'}], 'n1': [{'id': 'NCBIGene:7422'}], 'n2': [{'id': 'HP:0002099'}]}, 'normalized_score': 100.0, 'reasoner_id': 'BTE', 'score': 2.9068059062475973}, {'edge_bindings': {'e0': [{'id': 'a823bd04f08a5591fba87dc56d52647d'}], 'e1': [{'id': '7ee912dbb014b53ed3fcea0e7dc883f8'}, {'id': '651e711baac3604af1b09d0e92751d2d'}, {'id': '81900675077f00cbc42b8d26d95589fc'}]}, 'node_bindings': {'n0': [{'id': 'PUBCHEM.COMPOUND:123596'}], 'n1': [{'id': 'NCBIGene:841'}], 'n2': [{'id': 'HP:0002099'}]}, 'normalized_score': 85.71428571428571, 'reasoner_id': 'BTE', 'score': 1.1203189146560173}, {'edge_bindings': {'e0': [{'id': 'c2323af968047313b3f7d025c3f7c804'}, {'id': '3b84553e4e9b9466e82c594cc5913b49'}], 'e1': [{'id': '1a199ceb3b74961d7b72e222ac9b3172'}]}, 'node_bindings': {'n0': [{'id': 'PUBCHEM.COMPOUND:123596'}], 'n1': [{'id': 'NCBIGene:4067'}], 'n2': [{'id': 'HP:0002099'}]}, 'normalized_score': 71.42857142857143, 'reasoner_id': 'BTE', 'score': 0.8510301411652301}, {'edge_bindings': {'e0': [{'id': '02217ffa55351b0c3c0a32431158ca27'}], 'e1': [{'id': 'e2e73e88ecdce75377ff36aae8c52281'}, {'id': 'e0cd55dd7bf8840506b2dab758bfcd4e'}, {'id': 'ac85a6ff34dc2170e13fe0019d983c85'}]}, 'node_bindings': {'n0': [{'id': 'PUBCHEM.COMPOUND:123596'}], 'n1': [{'id': 'NCBIGene:3576'}], 'n2': [{'id': 'HP:0002099'}]}, 'normalized_score': 57.142857142857146, 'reasoner_id': 'BTE', 'score': 0.7966657259581773}, {'edge_bindings': {'e0': [{'id': '6ba3d9ff480ec75e66a38889f0e9c4b2'}], 'e1': [{'id': '8b7bd93b93c73afa7122d4b6f11c9258'}]}, 'node_bindings': {'n0': [{'id': 'PUBCHEM.COMPOUND:123596'}], 'n1': [{'id': 'NCBIGene:142'}], 'n2': [{'id': 'HP:0002099'}]}, 'normalized_score': 28.571428571428573, 'reasoner_id': 'BTE', 'score': 0}, {'edge_bindings': {'e0': [{'id': 'bca736d8e683b35a133bee44f2a75629'}, {'id': '83f80061137b13dfde6941f711ab53eb'}, {'id': '05ec582ef37cc52970d0549edb0da775'}], 'e1': [{'id': 'fbb3eab04ddeb56570d39d1ee2f82da3'}]}, 'node_bindings': {'n0': [{'id': 'PUBCHEM.COMPOUND:123596'}], 'n1': [{'id': 'NCBIGene:5156'}], 'n2': [{'id': 'HP:0002099'}]}, 'normalized_score': 28.571428571428573, 'reasoner_id': 'BTE', 'score': 0}, {'edge_bindings': {'e0': [{'id': '47b528f4619fe13a988467e04b78b536'}], 'e1': [{'id': '17a4c659b4a9f1a8bcd164cd17c03db6'}]}, 'node_bindings': {'n0': [{'id': 'PUBCHEM.COMPOUND:123596'}], 'n1': [{'id': 'NCBIGene:3725'}], 'n2': [{'id': 'HP:0002099'}]}, 'normalized_score': 28.571428571428573, 'reasoner_id': 'BTE', 'score': 0}]}, 'validation_result': {'message': '', 'n_edges': 24, 'n_nodes': 9, 'provenance_summary': {'n_sources': 13, 'predicate_counts': {'biolink:associated_with': 3, 'biolink:contributes_to': 2, 'biolink:decreases_expression_of': 1, 'biolink:decreases_secretion_of': 1, 'biolink:gene_associated_with_condition': 5, 'biolink:increases_activity_of': 2, 'biolink:increases_degradation_of': 1, 'biolink:increases_secretion_of': 2, 'biolink:occurs_together_in_literature_with': 3, 'biolink:response_affected_by': 2, 'biolink:response_decreased_by': 1, 'biolink:response_increased_by': 1}, 'provenance_counts': {\"biolink:associated_with --> biolink:aggregator_knowledge_source = ['infores:biothings-explorer']\": ['biolink:associated_with', 'biolink:aggregator_knowledge_source', ['infores:biothings-explorer'], 3], \"biolink:associated_with --> biolink:aggregator_knowledge_source = ['infores:biothings-semmeddb']\": ['biolink:associated_with', 'biolink:aggregator_knowledge_source', ['infores:biothings-semmeddb'], 3], \"biolink:associated_with --> biolink:primary_knowledge_source = ['infores:semmeddb']\": ['biolink:associated_with', 'biolink:primary_knowledge_source', ['infores:semmeddb'], 3], \"biolink:contributes_to --> biolink:aggregator_knowledge_source = ['infores:biothings-explorer']\": ['biolink:contributes_to', 'biolink:aggregator_knowledge_source', ['infores:biothings-explorer'], 2], 'biolink:contributes_to --> biolink:original_knowledge_source = infores:text-mining-provider-targeted': ['biolink:contributes_to', 'biolink:original_knowledge_source', 'infores:text-mining-provider-targeted', 2], 'biolink:contributes_to --> biolink:supporting_data_source = infores:pubmed': ['biolink:contributes_to', 'biolink:supporting_data_source', 'infores:pubmed', 2], \"biolink:decreases_expression_of --> biolink:aggregator_knowledge_source = ['infores:automat-ctd']\": ['biolink:decreases_expression_of', 'biolink:aggregator_knowledge_source', ['infores:automat-ctd'], 1], \"biolink:decreases_expression_of --> biolink:aggregator_knowledge_source = ['infores:biothings-explorer']\": ['biolink:decreases_expression_of', 'biolink:aggregator_knowledge_source', ['infores:biothings-explorer'], 1], \"biolink:decreases_expression_of --> biolink:original_knowledge_source = ['infores:ctd']\": ['biolink:decreases_expression_of', 'biolink:original_knowledge_source', ['infores:ctd'], 1], \"biolink:decreases_secretion_of --> biolink:aggregator_knowledge_source = ['infores:automat-ctd']\": ['biolink:decreases_secretion_of', 'biolink:aggregator_knowledge_source', ['infores:automat-ctd'], 1], \"biolink:decreases_secretion_of --> biolink:aggregator_knowledge_source = ['infores:biothings-explorer']\": ['biolink:decreases_secretion_of', 'biolink:aggregator_knowledge_source', ['infores:biothings-explorer'], 1], \"biolink:decreases_secretion_of --> biolink:original_knowledge_source = ['infores:ctd']\": ['biolink:decreases_secretion_of', 'biolink:original_knowledge_source', ['infores:ctd'], 1], \"biolink:gene_associated_with_condition --> biolink:aggregator_knowledge_source = ['infores:biolink-api']\": ['biolink:gene_associated_with_condition', 'biolink:aggregator_knowledge_source', ['infores:biolink-api'], 2], \"biolink:gene_associated_with_condition --> biolink:aggregator_knowledge_source = ['infores:biothings-explorer']\": ['biolink:gene_associated_with_condition', 'biolink:aggregator_knowledge_source', ['infores:biothings-explorer'], 5], \"biolink:gene_associated_with_condition --> biolink:aggregator_knowledge_source = ['infores:mydisease-info']\": ['biolink:gene_associated_with_condition', 'biolink:aggregator_knowledge_source', ['infores:mydisease-info'], 3], \"biolink:gene_associated_with_condition --> biolink:primary_knowledge_source = ['infores:disgenet']\": ['biolink:gene_associated_with_condition', 'biolink:primary_knowledge_source', ['infores:disgenet'], 3], \"biolink:gene_associated_with_condition --> biolink:primary_knowledge_source = ['infores:monarchinitiative']\": ['biolink:gene_associated_with_condition', 'biolink:primary_knowledge_source', ['infores:monarchinitiative'], 2], \"biolink:increases_activity_of --> biolink:aggregator_knowledge_source = ['infores:automat-ctd']\": ['biolink:increases_activity_of', 'biolink:aggregator_knowledge_source', ['infores:automat-ctd'], 2], \"biolink:increases_activity_of --> biolink:aggregator_knowledge_source = ['infores:biothings-explorer']\": ['biolink:increases_activity_of', 'biolink:aggregator_knowledge_source', ['infores:biothings-explorer'], 2], \"biolink:increases_activity_of --> biolink:original_knowledge_source = ['infores:ctd']\": ['biolink:increases_activity_of', 'biolink:original_knowledge_source', ['infores:ctd'], 2], \"biolink:increases_degradation_of --> biolink:aggregator_knowledge_source = ['infores:automat-ctd']\": ['biolink:increases_degradation_of', 'biolink:aggregator_knowledge_source', ['infores:automat-ctd'], 1], \"biolink:increases_degradation_of --> biolink:aggregator_knowledge_source = ['infores:biothings-explorer']\": ['biolink:increases_degradation_of', 'biolink:aggregator_knowledge_source', ['infores:biothings-explorer'], 1], \"biolink:increases_degradation_of --> biolink:original_knowledge_source = ['infores:ctd']\": ['biolink:increases_degradation_of', 'biolink:original_knowledge_source', ['infores:ctd'], 1], \"biolink:increases_secretion_of --> biolink:aggregator_knowledge_source = ['infores:automat-ctd']\": ['biolink:increases_secretion_of', 'biolink:aggregator_knowledge_source', ['infores:automat-ctd'], 2], \"biolink:increases_secretion_of --> biolink:aggregator_knowledge_source = ['infores:biothings-explorer']\": ['biolink:increases_secretion_of', 'biolink:aggregator_knowledge_source', ['infores:biothings-explorer'], 2], \"biolink:increases_secretion_of --> biolink:original_knowledge_source = ['infores:ctd']\": ['biolink:increases_secretion_of', 'biolink:original_knowledge_source', ['infores:ctd'], 2], \"biolink:occurs_together_in_literature_with --> biolink:aggregator_knowledge_source = ['infores:biothings-explorer']\": ['biolink:occurs_together_in_literature_with', 'biolink:aggregator_knowledge_source', ['infores:biothings-explorer'], 3], \"biolink:occurs_together_in_literature_with --> biolink:aggregator_knowledge_source = ['infores:biothings-pfocr']\": ['biolink:occurs_together_in_literature_with', 'biolink:aggregator_knowledge_source', ['infores:biothings-pfocr'], 3], \"biolink:occurs_together_in_literature_with --> biolink:primary_knowledge_source = ['infores:pfocr']\": ['biolink:occurs_together_in_literature_with', 'biolink:primary_knowledge_source', ['infores:pfocr'], 3], \"biolink:response_affected_by --> biolink:aggregator_knowledge_source = ['infores:automat-ctd']\": ['biolink:response_affected_by', 'biolink:aggregator_knowledge_source', ['infores:automat-ctd'], 2], \"biolink:response_affected_by --> biolink:aggregator_knowledge_source = ['infores:biothings-explorer']\": ['biolink:response_affected_by', 'biolink:aggregator_knowledge_source', ['infores:biothings-explorer'], 2], \"biolink:response_affected_by --> biolink:original_knowledge_source = ['infores:ctd']\": ['biolink:response_affected_by', 'biolink:original_knowledge_source', ['infores:ctd'], 2], \"biolink:response_decreased_by --> biolink:aggregator_knowledge_source = ['infores:automat-ctd']\": ['biolink:response_decreased_by', 'biolink:aggregator_knowledge_source', ['infores:automat-ctd'], 1], \"biolink:response_decreased_by --> biolink:aggregator_knowledge_source = ['infores:biothings-explorer']\": ['biolink:response_decreased_by', 'biolink:aggregator_knowledge_source', ['infores:biothings-explorer'], 1], \"biolink:response_decreased_by --> biolink:original_knowledge_source = ['infores:ctd']\": ['biolink:response_decreased_by', 'biolink:original_knowledge_source', ['infores:ctd'], 1], \"biolink:response_increased_by --> biolink:aggregator_knowledge_source = ['infores:automat-ctd']\": ['biolink:response_increased_by', 'biolink:aggregator_knowledge_source', ['infores:automat-ctd'], 1], \"biolink:response_increased_by --> biolink:aggregator_knowledge_source = ['infores:biothings-explorer']\": ['biolink:response_increased_by', 'biolink:aggregator_knowledge_source', ['infores:biothings-explorer'], 1], \"biolink:response_increased_by --> biolink:original_knowledge_source = ['infores:ctd']\": ['biolink:response_increased_by', 'biolink:original_knowledge_source', ['infores:ctd'], 1]}}, 'size': '83 kB', 'status': 'PASS', 'version': '1.3.0'}, 'workflow': [{'id': 'lookup'}]}\n"
     ]
    }
   ],
   "source": [
    "# store the URL in url as parameter for urlopen\n",
    "trapi_results_url = \"https://arax.ncats.io/api/arax/v1.3/response/7b14f961-9066-41f7-9e3b-d76b2b4a7fac\"\n",
    "   \n",
    "# storing the JSON response from url\n",
    "with urlopen(trapi_results_url) as url:\n",
    "    trapi_results = json.load(url)\n",
    "\n",
    "# print the json response\n",
    "print(trapi_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'knowledge_graph': {'edges': {'02217ffa55351b0c3c0a32431158ca27': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:increases_secretion_of', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:3576', 'predicate': 'biolink:increases_secretion_of', 'subject': 'PUBCHEM.COMPOUND:123596'}, '05ec582ef37cc52970d0549edb0da775': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:increases_response_to', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:5156', 'predicate': 'biolink:response_affected_by', 'subject': 'PUBCHEM.COMPOUND:123596'}, '133406fa1089f0630ad505ba930f0615': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:disgenet'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:mydisease-info'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:publications', 'value': ['PMID:24152156']}, {'attribute_type_id': 'disgenet-DPI', 'value': 0.923}, {'attribute_type_id': 'disgenet-DSI', 'value': 0.266}, {'attribute_type_id': 'disgenet-EI', 'value': 1}, {'attribute_type_id': 'disgenet-score', 'value': 0.01}, {'attribute_type_id': 'disgenet_source', 'value': 'BEFREE'}], 'object': 'HP:0002099', 'predicate': 'biolink:gene_associated_with_condition', 'subject': 'NCBIGene:7422'}, '17a4c659b4a9f1a8bcd164cd17c03db6': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:pfocr'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-pfocr'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'figure_download_url', 'value': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4973429/bin/fphar-07-00237-g0002.jpg'}, {'attribute_type_id': 'figure_title', 'value': 'Involvement of AKR1B1 in pathophysiology'}, {'attribute_type_id': 'pmc_reference', 'value': 'PMC4973429'}], 'object': 'HP:0002099', 'predicate': 'biolink:occurs_together_in_literature_with', 'subject': 'NCBIGene:3725'}, '1a199ceb3b74961d7b72e222ac9b3172': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:semmeddb'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-semmeddb'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:publications', 'value': ['PMID:28024734']}, {'attribute_type_id': 'biolink:original_object', 'value': 'C0004096'}, {'attribute_type_id': 'biolink:original_predicate', 'value': 'ASSOCIATED_WITH'}, {'attribute_type_id': 'biolink:original_subject', 'value': 'C0812307'}, {'attribute_type_id': 'original_object_name', 'value': 'Asthma'}, {'attribute_type_id': 'original_subject_name', 'value': 'LYN gene'}], 'object': 'HP:0002099', 'predicate': 'biolink:associated_with', 'subject': 'NCBIGene:4067'}, '23a8bbe1078a780814d37f2785fad1ab': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:increases_secretion_of', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:7422', 'predicate': 'biolink:increases_secretion_of', 'subject': 'PUBCHEM.COMPOUND:123596'}, '3b84553e4e9b9466e82c594cc5913b49': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:decreases_response_to', 'value_type_id': 'EDAM:data_0006', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}], 'object': 'NCBIGene:4067', 'predicate': 'biolink:response_decreased_by', 'subject': 'PUBCHEM.COMPOUND:123596'}, '3bdac66d7282f313bc444533ecb84a4f': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:original_knowledge_source', 'value': 'infores:text-mining-provider-targeted', 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_data_source', 'value': 'infores:pubmed', 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:has_evidence_count', 'value': 3, 'value_type_id': 'biolink:EvidenceCount'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:tmkp_confidence_score', 'value': 0.99767165, 'value_type_id': 'biolink:ConfidenceLevel'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document', 'value': 'PMID:19196819|PMID:19196819|PMID:19196819', 'value_type_id': 'biolink:Publication'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_study_result', 'attributes': [{'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_text', 'value': 'VEGF polymorphisms are associated with childhood asthma, lung function and airway responsiveness in two populations, suggesting that VEGF polymorphisms influence asthma susceptibility, airflow obstruction and airways responsiveness.', 'value_type_id': 'EDAM:data_3671'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document', 'value': 'PMID:19196819', 'value_type_id': 'biolink:Publication', 'value_url': 'https://pubmed.ncbi.nlm.nih.gov/19196819/'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_text_located_in', 'value': 'abstract', 'value_type_id': 'IAO_0000314'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:extraction_confidence_score', 'value': 0.9979881, 'value_type_id': 'EDAM:data_1772'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:subject_location_in_text', 'value': '0|4', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:text-mining-provider-targeted ', 'attribute_type_id': 'biolink:object_location_in_text', 'value': '49|55', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document_year', 'value': 2009, 'value_type_id': 'UO:0000036'}], 'value': 'tmkp:0e7972a590450fdf5d10adaa72434ef7fa501362e9d172499c1735d77e13ad2d', 'value_type_id': 'biolink:TextMiningResult'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_study_result', 'attributes': [{'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_text', 'value': 'VEGF polymorphisms are associated with childhood asthma, lung function and airway responsiveness in two populations, suggesting that VEGF polymorphisms influence asthma susceptibility, airflow obstruction and airways responsiveness.', 'value_type_id': 'EDAM:data_3671'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document', 'value': 'PMID:19196819', 'value_type_id': 'biolink:Publication', 'value_url': 'https://pubmed.ncbi.nlm.nih.gov/19196819/'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_text_located_in', 'value': 'abstract', 'value_type_id': 'IAO_0000314'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:extraction_confidence_score', 'value': 0.9977518, 'value_type_id': 'EDAM:data_1772'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:subject_location_in_text', 'value': '133|137', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:text-mining-provider-targeted ', 'attribute_type_id': 'biolink:object_location_in_text', 'value': '49|55', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document_year', 'value': 2009, 'value_type_id': 'UO:0000036'}], 'value': 'tmkp:338c6421377e7f6ef40197260064a6d7015bde00a8cb774fd97c3c385b9e9904', 'value_type_id': 'biolink:TextMiningResult'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_study_result', 'attributes': [{'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_text', 'value': 'Association of VEGF polymorphisms with childhood asthma, lung function and airway responsiveness.', 'value_type_id': 'EDAM:data_3671'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document', 'value': 'PMID:19196819', 'value_type_id': 'biolink:Publication', 'value_url': 'https://pubmed.ncbi.nlm.nih.gov/19196819/'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_text_located_in', 'value': 'title', 'value_type_id': 'IAO_0000314'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:extraction_confidence_score', 'value': 0.99727505, 'value_type_id': 'EDAM:data_1772'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:subject_location_in_text', 'value': '15|19', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:text-mining-provider-targeted ', 'attribute_type_id': 'biolink:object_location_in_text', 'value': '49|55', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document_year', 'value': 2009, 'value_type_id': 'UO:0000036'}], 'value': 'tmkp:10fd1e6a9b3c224a0813d96e4e928c65327c82b2385164c4cf7c82456c21b097', 'value_type_id': 'biolink:TextMiningResult'}], 'object': 'HP:0002099', 'predicate': 'biolink:contributes_to', 'subject': 'NCBIGene:7422'}, '47b528f4619fe13a988467e04b78b536': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:increases_activity_of', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:3725', 'predicate': 'biolink:increases_activity_of', 'subject': 'PUBCHEM.COMPOUND:123596'}, '651e711baac3604af1b09d0e92751d2d': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:disgenet'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:mydisease-info'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:publications', 'value': ['PMID:18823309']}, {'attribute_type_id': 'disgenet-DPI', 'value': 0.923}, {'attribute_type_id': 'disgenet-DSI', 'value': 0.404}, {'attribute_type_id': 'disgenet-EI', 'value': 1}, {'attribute_type_id': 'disgenet-score', 'value': 0.01}, {'attribute_type_id': 'disgenet_source', 'value': 'BEFREE'}], 'object': 'HP:0002099', 'predicate': 'biolink:gene_associated_with_condition', 'subject': 'NCBIGene:841'}, '6ba3d9ff480ec75e66a38889f0e9c4b2': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:increases_degradation_of', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:142', 'predicate': 'biolink:increases_degradation_of', 'subject': 'PUBCHEM.COMPOUND:123596'}, '79c5be38756c7936c6706468d30a3133': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:decreases_secretion_of', 'value_type_id': 'EDAM:data_0006', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}], 'object': 'NCBIGene:7422', 'predicate': 'biolink:decreases_secretion_of', 'subject': 'PUBCHEM.COMPOUND:123596'}, '7ee912dbb014b53ed3fcea0e7dc883f8': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:monarchinitiative'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biolink-api'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:publications', 'value': ['PMID:12353035']}, {'attribute_type_id': 'biolink:original_predicate', 'value': 'has phenotype'}, {'attribute_type_id': 'monarch_source_database', 'value': ['https://archive.monarchinitiative.org/#orphanet', 'https://archive.monarchinitiative.org/#hpoa']}], 'object': 'HP:0002099', 'predicate': 'biolink:gene_associated_with_condition', 'subject': 'NCBIGene:841'}, '81900675077f00cbc42b8d26d95589fc': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:semmeddb'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-semmeddb'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:publications', 'value': ['PMID:18823309']}, {'attribute_type_id': 'biolink:original_object', 'value': 'C0004096'}, {'attribute_type_id': 'biolink:original_predicate', 'value': 'ASSOCIATED_WITH'}, {'attribute_type_id': 'biolink:original_subject', 'value': 'C1332667'}, {'attribute_type_id': 'original_object_name', 'value': 'Asthma'}, {'attribute_type_id': 'original_subject_name', 'value': 'CASP8 gene'}], 'object': 'HP:0002099', 'predicate': 'biolink:associated_with', 'subject': 'NCBIGene:841'}, '83f80061137b13dfde6941f711ab53eb': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:decreases_expression_of', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:5156', 'predicate': 'biolink:decreases_expression_of', 'subject': 'PUBCHEM.COMPOUND:123596'}, '8b7bd93b93c73afa7122d4b6f11c9258': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:pfocr'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-pfocr'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'figure_download_url', 'value': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8065335/bin/41392_2021_568_Fig5_HTML.jpg'}, {'attribute_type_id': 'figure_title', 'value': 'Putative mechanisms for CVL218 as a PARP1 inhibitor to combat the COVID-19 disease, derived based on the data present in this study and the known antiviral activities of PARP1 inhibitors previously reported in the literature'}, {'attribute_type_id': 'pmc_reference', 'value': 'PMC8065335'}], 'object': 'HP:0002099', 'predicate': 'biolink:occurs_together_in_literature_with', 'subject': 'NCBIGene:142'}, 'a823bd04f08a5591fba87dc56d52647d': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:increases_activity_of', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:841', 'predicate': 'biolink:increases_activity_of', 'subject': 'PUBCHEM.COMPOUND:123596'}, 'ac85a6ff34dc2170e13fe0019d983c85': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:pfocr'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-pfocr'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'figure_download_url', 'value': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8065335/bin/41392_2021_568_Fig5_HTML.jpg'}, {'attribute_type_id': 'figure_title', 'value': 'Putative mechanisms for CVL218 as a PARP1 inhibitor to combat the COVID-19 disease, derived based on the data present in this study and the known antiviral activities of PARP1 inhibitors previously reported in the literature'}, {'attribute_type_id': 'pmc_reference', 'value': 'PMC8065335'}], 'object': 'HP:0002099', 'predicate': 'biolink:occurs_together_in_literature_with', 'subject': 'NCBIGene:3576'}, 'bca736d8e683b35a133bee44f2a75629': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:increases_response_to', 'value_type_id': 'EDAM:data_0006', 'value_url': None}], 'object': 'NCBIGene:5156', 'predicate': 'biolink:response_increased_by', 'subject': 'PUBCHEM.COMPOUND:123596'}, 'c2323af968047313b3f7d025c3f7c804': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:original_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:original_knowledge_source', 'value': ['infores:ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}, {'attribute_source': None, 'attribute_type_id': 'biolink:relation', 'attributes': None, 'description': None, 'original_attribute_name': 'relation', 'value': 'CTD:decreases_response_to', 'value_type_id': 'EDAM:data_0006', 'value_url': None}, {'attribute_source': 'infores:automat-ctd', 'attribute_type_id': 'biolink:aggregator_knowledge_source', 'attributes': None, 'description': None, 'original_attribute_name': 'biolink:aggregator_knowledge_source', 'value': ['infores:automat-ctd'], 'value_type_id': 'biolink:InformationResource', 'value_url': None}], 'object': 'NCBIGene:4067', 'predicate': 'biolink:response_affected_by', 'subject': 'PUBCHEM.COMPOUND:123596'}, 'c80d1f9d65f26b19dec2ca1395d1988e': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:semmeddb'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-semmeddb'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:publications', 'value': ['PMID:18421428']}, {'attribute_type_id': 'biolink:original_object', 'value': 'C0004096'}, {'attribute_type_id': 'biolink:original_predicate', 'value': 'ASSOCIATED_WITH'}, {'attribute_type_id': 'biolink:original_subject', 'value': 'C1171892'}, {'attribute_type_id': 'original_object_name', 'value': 'Asthma'}, {'attribute_type_id': 'original_subject_name', 'value': 'VEGF protein, human'}], 'object': 'HP:0002099', 'predicate': 'biolink:associated_with', 'subject': 'NCBIGene:7422'}, 'e0cd55dd7bf8840506b2dab758bfcd4e': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:original_knowledge_source', 'value': 'infores:text-mining-provider-targeted', 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_data_source', 'value': 'infores:pubmed', 'value_type_id': 'biolink:InformationResource'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:has_evidence_count', 'value': 1, 'value_type_id': 'biolink:EvidenceCount'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:tmkp_confidence_score', 'value': 0.9979808, 'value_type_id': 'biolink:ConfidenceLevel'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document', 'value': 'PMID:25530067', 'value_type_id': 'biolink:Publication'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_study_result', 'attributes': [{'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:supporting_text', 'value': 'Compared with patients with asthma attacks, patients with exacerbated COPD had increased IL-8 expression in both serum and EBC.', 'value_type_id': 'EDAM:data_3671'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document', 'value': 'PMID:25530067', 'value_type_id': 'biolink:Publication', 'value_url': 'https://pubmed.ncbi.nlm.nih.gov/25530067/'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_text_located_in', 'value': 'abstract', 'value_type_id': 'IAO_0000314'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:extraction_confidence_score', 'value': 0.9979808, 'value_type_id': 'EDAM:data_1772'}, {'attribute_source': 'infores:text-mining-provider-targeted', 'attribute_type_id': 'biolink:subject_location_in_text', 'value': '89|93', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:text-mining-provider-targeted ', 'attribute_type_id': 'biolink:object_location_in_text', 'value': '28|34', 'value_type_id': 'SIO:001056'}, {'attribute_source': 'infores:pubmed', 'attribute_type_id': 'biolink:supporting_document_year', 'value': 2014, 'value_type_id': 'UO:0000036'}], 'value': 'tmkp:2e332a4f0a754fd9087e3730db021775f5b4815d37a79ab2b3d5532fc6f5b828', 'value_type_id': 'biolink:TextMiningResult'}], 'object': 'HP:0002099', 'predicate': 'biolink:contributes_to', 'subject': 'NCBIGene:3576'}, 'e2e73e88ecdce75377ff36aae8c52281': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:disgenet'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:mydisease-info'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:publications', 'value': ['PMID:29189668']}, {'attribute_type_id': 'disgenet-DPI', 'value': 0.962}, {'attribute_type_id': 'disgenet-DSI', 'value': 0.31}, {'attribute_type_id': 'disgenet-EI', 'value': 1}, {'attribute_type_id': 'disgenet-score', 'value': 0.01}, {'attribute_type_id': 'disgenet_source', 'value': 'BEFREE'}], 'object': 'HP:0002099', 'predicate': 'biolink:gene_associated_with_condition', 'subject': 'NCBIGene:3576'}, 'fbb3eab04ddeb56570d39d1ee2f82da3': {'attributes': [{'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biothings-explorer'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:primary_knowledge_source', 'value': ['infores:monarchinitiative'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:aggregator_knowledge_source', 'value': ['infores:biolink-api'], 'value_type_id': 'biolink:InformationResource'}, {'attribute_type_id': 'biolink:original_predicate', 'value': 'has phenotype'}, {'attribute_type_id': 'monarch_source_database', 'value': ['https://archive.monarchinitiative.org/#omim', 'https://archive.monarchinitiative.org/#hpoa']}], 'object': 'HP:0002099', 'predicate': 'biolink:gene_associated_with_condition', 'subject': 'NCBIGene:5156'}}, 'nodes': {'HP:0002099': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['HP:0002099', 'NCIT:C113673', 'UMLS:C0004096', 'UMLS:C0085129', 'UMLS:C0340062', 'UMLS:C3714497', 'MEDDRA:10003553', 'MEDDRA:10003555', 'MEDDRA:10003560', 'MEDDRA:10003561', 'MEDDRA:10003565', 'MEDDRA:10006450', 'MEDDRA:10037993', 'MEDDRA:10056285', 'MEDDRA:10066091', 'MEDDRA:10066863', 'SNOMEDCT:195967001', 'SNOMEDCT:991000119106', 'MESH:D001249', 'MESH:D016535']}, {'attribute_type_id': 'biolink:synonym', 'value': ['Asthma', 'Reactive Airway Disease', 'Bronchial Hyperreactivity', 'Hyperreactive airway disease', 'Reactive airway disease']}], 'categories': ['biolink:PhenotypicFeature'], 'name': 'Asthma'}, 'NCBIGene:142': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['NCBIGene:142', 'ENSEMBL:ENSG00000143799', 'HGNC:270', 'OMIM:173870', 'UMLS:C1538577', 'UniProtKB:A0A024R3T8', 'UniProtKB:P09874', 'PR:P09874', 'ENSEMBL:ENSP00000355759', 'UMLS:C1567710', 'UMLS:C3641724']}, {'attribute_type_id': 'biolink:synonym', 'value': ['PARP1', 'PARP1 gene', 'A0A024R3T8_HUMAN Poly [ADP-ribose] polymerase (trembl)', 'PARP1_HUMAN Poly [ADP-ribose] polymerase 1 (sprot)', 'poly [ADP-ribose] polymerase 1 (human)', 'PARP1 protein, human', 'Poly [ADP-Ribose] Polymerase 1 p85, human']}], 'categories': ['biolink:Gene'], 'name': 'PARP1'}, 'NCBIGene:3576': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['NCBIGene:3576', 'ENSEMBL:ENSG00000169429', 'HGNC:6025', 'OMIM:146930', 'UMLS:C1366571', 'UniProtKB:A0A024RDA5', 'UniProtKB:P10145', 'PR:P10145', 'ENSEMBL:ENSP00000306512', 'UMLS:C1698756']}, {'attribute_type_id': 'biolink:synonym', 'value': ['CXCL8', 'CXCL8 gene', 'A0A024RDA5_HUMAN Multifunctional fusion protein (trembl)', 'IL8_HUMAN Interleukin-8 (sprot)', 'interleukin-8 (human)', 'CXCL8 protein, human']}], 'categories': ['biolink:Gene'], 'name': 'CXCL8'}, 'NCBIGene:3725': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['NCBIGene:3725', 'ENSEMBL:ENSG00000177606', 'HGNC:6204', 'OMIM:165160', 'UMLS:C0162508', 'UniProtKB:P05412', 'PR:P05412', 'ENSEMBL:ENSP00000360266', 'ENSEMBL:ENSP00000503132', 'UMLS:C3536909']}, {'attribute_type_id': 'biolink:synonym', 'value': ['JUN', 'JUN gene', 'JUN_HUMAN Transcription factor AP-1 (sprot)', 'transcription factor AP-1 (human)', 'Transcription Factor AP-1, human']}], 'categories': ['biolink:Gene'], 'name': 'JUN'}, 'NCBIGene:4067': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['NCBIGene:4067', 'ENSEMBL:ENSG00000254087', 'HGNC:6735', 'OMIM:165120', 'UMLS:C0812307', 'UniProtKB:A8K379', 'UniProtKB:P07948', 'PR:P07948', 'UMLS:C1442804', 'UniProtKB:Q6NUK7']}, {'attribute_type_id': 'biolink:synonym', 'value': ['LYN', 'LYN gene', 'A8K379_HUMAN Tyrosine-protein kinase (trembl)', 'LYN_HUMAN Tyrosine-protein kinase Lyn (sprot)', 'tyrosine-protein kinase Lyn (human)', 'LYN protein, human', 'Q6NUK7_HUMAN Tyrosine-protein kinase (Fragment) (trembl)']}], 'categories': ['biolink:Gene'], 'name': 'LYN'}, 'NCBIGene:5156': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['NCBIGene:5156', 'ENSEMBL:ENSG00000134853', 'HGNC:8803', 'OMIM:173490', 'UMLS:C1335201', 'UniProtKB:P16234', 'PR:P16234', 'UMLS:C3853694']}, {'attribute_type_id': 'biolink:synonym', 'value': ['PDGFRA', 'PDGFRA gene', 'PGFRA_HUMAN Platelet-derived growth factor receptor alpha (sprot)', 'platelet-derived growth factor receptor alpha (human)', 'Platelet-Derived Growth Factor Receptor Alpha, human']}], 'categories': ['biolink:Gene'], 'name': 'PDGFRA'}, 'NCBIGene:7422': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['NCBIGene:7422', 'ENSEMBL:ENSG00000112715', 'HGNC:12680', 'OMIM:192240', 'UMLS:C1823619', 'UniProtKB:A0A024RD33', 'UniProtKB:A0A024RD37', 'UniProtKB:A0A0Y0IMM4', 'UniProtKB:A2A2V4', 'ENSEMBL:ENSP00000230480', 'UniProtKB:P15692', 'PR:P15692', 'UMLS:C0376854', 'UMLS:C0376855', 'UMLS:C0376856', 'UMLS:C0525171', 'UMLS:C0752432', 'UMLS:C1171892']}, {'attribute_type_id': 'biolink:synonym', 'value': ['VEGFA', 'VEGFA gene', 'A0A024RD33_HUMAN Vascular endothelial growth factor A (trembl)', 'A0A024RD37_HUMAN Vascular endothelial growth factor A (trembl)', 'A0A0Y0IMM4_HUMAN Vascular endothelial growth factor A (Fragment) (trembl)', 'A2A2V4_HUMAN Vascular endothelial growth factor A (trembl)', 'VEGFA_HUMAN Vascular endothelial growth factor A (sprot)', 'vascular endothelial growth factor A (human)', 'VEGF165 protein, human', 'VEGF189 protein, human', 'VEGF206 protein, human', 'VEGF145 protein, human', 'VEGF183 protein, human', 'VEGF protein, human']}], 'categories': ['biolink:Gene'], 'name': 'VEGFA'}, 'NCBIGene:841': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['NCBIGene:841', 'ENSEMBL:ENSG00000064012', 'HGNC:1509', 'OMIM:601763', 'UMLS:C1332667', 'UniProtKB:A0A024R3Z8', 'UniProtKB:Q14790', 'PR:Q14790', 'UMLS:C1144061']}, {'attribute_type_id': 'biolink:synonym', 'value': ['CASP8', 'CASP8 gene', 'A0A024R3Z8_HUMAN HCG16983, isoform CRA_b (trembl)', 'CASP8_HUMAN Caspase-8 (sprot)', 'caspase-8 (human)', 'CASP8 protein, human']}], 'categories': ['biolink:Gene'], 'name': 'CASP8'}, 'PUBCHEM.COMPOUND:123596': {'attributes': [{'attribute_type_id': 'biolink:xref', 'value': ['PUBCHEM.COMPOUND:123596', 'PUBCHEM.COMPOUND:90091706', 'CHEMBL.COMPOUND:CHEMBL1642', 'UNII:8A1O1M485B', 'CHEBI:31690', 'MESH:D000068877', 'CAS:220127-57-1', 'INCHIKEY:YLMAHDNUQAMNNX-UHFFFAOYSA-N', 'UMLS:C0906802', 'UMLS:C0939537']}, {'attribute_type_id': 'biolink:synonym', 'value': ['Imatinib mesylate', 'Methanesulfonate;4-[(4-methylpiperazin-4-ium-1-yl)methyl]-N-[4-methyl-3-[(4-pyridin-3-ylpyrimidin-2-yl)amino]phenyl]benzamide', 'IMATINIB MESYLATE', 'imatinib methanesulfonate', 'Imatinib Mesylate', 'STI571']}], 'categories': ['biolink:MolecularMixture'], 'name': 'Imatinib mesylate'}}}, 'query_graph': {'edges': {'e0': {'object': 'n1', 'predicates': ['biolink:related_to'], 'subject': 'n0'}, 'e1': {'object': 'n2', 'predicates': ['biolink:related_to'], 'subject': 'n1'}}, 'nodes': {'n0': {'categories': ['biolink:Drug', 'biolink:MolecularMixture'], 'ids': ['MESH:D000068877']}, 'n1': {'categories': ['biolink:Gene']}, 'n2': {'categories': ['biolink:Disease', 'biolink:PhenotypicFeature'], 'ids': ['MESH:D001249']}}}, 'results': [{'edge_bindings': {'e0': [{'id': '23a8bbe1078a780814d37f2785fad1ab'}, {'id': '79c5be38756c7936c6706468d30a3133'}], 'e1': [{'id': '133406fa1089f0630ad505ba930f0615'}, {'id': '3bdac66d7282f313bc444533ecb84a4f'}, {'id': 'c80d1f9d65f26b19dec2ca1395d1988e'}]}, 'node_bindings': {'n0': [{'id': 'PUBCHEM.COMPOUND:123596'}], 'n1': [{'id': 'NCBIGene:7422'}], 'n2': [{'id': 'HP:0002099'}]}, 'normalized_score': 100.0, 'reasoner_id': 'BTE', 'score': 2.9068059062475973}, {'edge_bindings': {'e0': [{'id': 'a823bd04f08a5591fba87dc56d52647d'}], 'e1': [{'id': '7ee912dbb014b53ed3fcea0e7dc883f8'}, {'id': '651e711baac3604af1b09d0e92751d2d'}, {'id': '81900675077f00cbc42b8d26d95589fc'}]}, 'node_bindings': {'n0': [{'id': 'PUBCHEM.COMPOUND:123596'}], 'n1': [{'id': 'NCBIGene:841'}], 'n2': [{'id': 'HP:0002099'}]}, 'normalized_score': 85.71428571428571, 'reasoner_id': 'BTE', 'score': 1.1203189146560173}, {'edge_bindings': {'e0': [{'id': 'c2323af968047313b3f7d025c3f7c804'}, {'id': '3b84553e4e9b9466e82c594cc5913b49'}], 'e1': [{'id': '1a199ceb3b74961d7b72e222ac9b3172'}]}, 'node_bindings': {'n0': [{'id': 'PUBCHEM.COMPOUND:123596'}], 'n1': [{'id': 'NCBIGene:4067'}], 'n2': [{'id': 'HP:0002099'}]}, 'normalized_score': 71.42857142857143, 'reasoner_id': 'BTE', 'score': 0.8510301411652301}, {'edge_bindings': {'e0': [{'id': '02217ffa55351b0c3c0a32431158ca27'}], 'e1': [{'id': 'e2e73e88ecdce75377ff36aae8c52281'}, {'id': 'e0cd55dd7bf8840506b2dab758bfcd4e'}, {'id': 'ac85a6ff34dc2170e13fe0019d983c85'}]}, 'node_bindings': {'n0': [{'id': 'PUBCHEM.COMPOUND:123596'}], 'n1': [{'id': 'NCBIGene:3576'}], 'n2': [{'id': 'HP:0002099'}]}, 'normalized_score': 57.142857142857146, 'reasoner_id': 'BTE', 'score': 0.7966657259581773}, {'edge_bindings': {'e0': [{'id': '6ba3d9ff480ec75e66a38889f0e9c4b2'}], 'e1': [{'id': '8b7bd93b93c73afa7122d4b6f11c9258'}]}, 'node_bindings': {'n0': [{'id': 'PUBCHEM.COMPOUND:123596'}], 'n1': [{'id': 'NCBIGene:142'}], 'n2': [{'id': 'HP:0002099'}]}, 'normalized_score': 28.571428571428573, 'reasoner_id': 'BTE', 'score': 0}, {'edge_bindings': {'e0': [{'id': 'bca736d8e683b35a133bee44f2a75629'}, {'id': '83f80061137b13dfde6941f711ab53eb'}, {'id': '05ec582ef37cc52970d0549edb0da775'}], 'e1': [{'id': 'fbb3eab04ddeb56570d39d1ee2f82da3'}]}, 'node_bindings': {'n0': [{'id': 'PUBCHEM.COMPOUND:123596'}], 'n1': [{'id': 'NCBIGene:5156'}], 'n2': [{'id': 'HP:0002099'}]}, 'normalized_score': 28.571428571428573, 'reasoner_id': 'BTE', 'score': 0}, {'edge_bindings': {'e0': [{'id': '47b528f4619fe13a988467e04b78b536'}], 'e1': [{'id': '17a4c659b4a9f1a8bcd164cd17c03db6'}]}, 'node_bindings': {'n0': [{'id': 'PUBCHEM.COMPOUND:123596'}], 'n1': [{'id': 'NCBIGene:3725'}], 'n2': [{'id': 'HP:0002099'}]}, 'normalized_score': 28.571428571428573, 'reasoner_id': 'BTE', 'score': 0}]}\n"
     ]
    }
   ],
   "source": [
    "# get the TRAPI results and knowledge graph from the jason response\n",
    "trapi_message = trapi_results['message']\n",
    "\n",
    "# print the TRAPI results and knowledge graph\n",
    "print(trapi_message)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a dataframe with all TRAPI results\n",
    "  \n",
    "The final dataframe is called ```trapi_results_df```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: we're including categories specified in the query template\n",
    "# in the next cell, and in a cell further down, we're also including\n",
    "# any categories from the TRAPI results.\n",
    "#\n",
    "# TODO: Some categories are supersets of others. Should we handle\n",
    "# this systematically?\n",
    "\n",
    "# get the query graph from the TRAPI results\n",
    "query = trapi_message[\"query_graph\"]\n",
    "\n",
    "curie_categories = set()\n",
    "for v in query[\"nodes\"].values():\n",
    "    if \"categories\" in v:\n",
    "        for category in v[\"categories\"]:\n",
    "            curie_categories.add(category)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all CURIE prefixes found:\n",
      "{'UMLS', 'HP', 'NCIT', 'OMIM', 'PUBCHEM.COMPOUND', 'CHEMBL.COMPOUND', 'HGNC', 'ENSEMBL', 'MESH', 'UniProtKB', 'SNOMEDCT', 'PR', 'CAS', 'CHEBI', 'INCHIKEY', 'NCBIGene', 'MEDDRA', 'UNII'}\n",
      "\n",
      "unified CURIE prefixes found:\n",
      "{'NCBIGene', 'MESH'}\n",
      "\n",
      "failed to unify 0 CURIEs\n"
     ]
    }
   ],
   "source": [
    "# for genes/gene products, chemicals and diseases\n",
    "# we need to resolve the CURIES as PFOCR data has only mesh ids and NCBI gene ids\n",
    "preferred_prefixes = set([\"NCBIGene\", \"MESH\"])\n",
    "trapi_results_unified_curies = set()\n",
    "unification_failed_curies = set()\n",
    "unified_prefixes = set()\n",
    "all_prefixes = set()\n",
    "curie_to_name = dict()\n",
    "curie_to_categories = dict()\n",
    "curie_to_unified_curie = dict()\n",
    "for k, v in trapi_message[\"knowledge_graph\"][\"nodes\"].items():\n",
    "    name = v[\"name\"]\n",
    "    categories = v[\"categories\"]\n",
    "    curie_categories |= set(categories)\n",
    "\n",
    "    for a in v[\"attributes\"]:\n",
    "        if a[\"attribute_type_id\"] == \"biolink:xref\":\n",
    "            curies = a[\"value\"]\n",
    "\n",
    "            # k should always be one of the curies\n",
    "            if not k in curies:\n",
    "                raise Exception(f\"key {k} not in {curies}\")\n",
    "\n",
    "            unified_curie = None\n",
    "            intersecting_trapi_results_unified_curies = trapi_results_unified_curies.intersection(\n",
    "                set(curies)\n",
    "            )\n",
    "            if len(intersecting_trapi_results_unified_curies) > 1:\n",
    "                multiple_matches = list(intersecting_trapi_results_unified_curies)\n",
    "                raise Exception(f\"matching multiple: {k} to {multiple_matches}\")\n",
    "            elif len(intersecting_trapi_results_unified_curies) == 1:\n",
    "                unified_curie = list(intersecting_trapi_results_unified_curies)[0]\n",
    "            else:\n",
    "                # get curie for preferred prefix. usually this is k, but not always.\n",
    "                for curie in curies:\n",
    "                    [prefix, identifier] = curie.split(\":\")\n",
    "                    if prefix in preferred_prefixes:\n",
    "                        unified_curie = curie\n",
    "                        trapi_results_unified_curies.add(unified_curie)\n",
    "                        break\n",
    "\n",
    "            if not unified_curie:\n",
    "                if k in curie_to_unified_curie:\n",
    "                    unified_curie = curie_to_unified_curie[k]\n",
    "                else:\n",
    "                    unification_failed_curies.add(k)\n",
    "                    break\n",
    "\n",
    "                #raise Exception(f\"failed to find a unified curie for {k} in {curies}\")\n",
    "\n",
    "            [unified_prefix, unified_identifier] = unified_curie.split(\":\")\n",
    "            unified_prefixes.add(unified_prefix)\n",
    "\n",
    "            for curie in curies:\n",
    "                [prefix, identifier] = curie.split(\":\")\n",
    "                all_prefixes.add(prefix)\n",
    "                if not curie in curie_to_unified_curie:\n",
    "                    curie_to_unified_curie[curie] = unified_curie\n",
    "                if not curie in curie_to_name:\n",
    "                    curie_to_name[curie] = name\n",
    "                    curie_to_categories[curie] = categories\n",
    "                elif curie_to_name[curie] != name:\n",
    "                    print(f\"curie {curie} has multiple primary names: {curie_to_name[curie]} and {name}\")\n",
    "                    #raise Exception(f\"curie{curie} has multiple names: {curie_to_name[curie]} and {name}\")\n",
    "\n",
    "\n",
    "#print(f'curie_categories: {curie_categories}')\n",
    "#print(\"\")\n",
    "#print(f\"curie_to_name key count: {len(curie_to_name.keys())}\")\n",
    "#print(f\"curie_to_unified_curie key count: {len(curie_to_unified_curie.keys())}\")\n",
    "#print(\"\")\n",
    "print(\"all CURIE prefixes found:\")\n",
    "print(all_prefixes)\n",
    "print(\"\")\n",
    "print(\"unified CURIE prefixes found:\")\n",
    "print(unified_prefixes)\n",
    "print(\"\")\n",
    "print(f\"failed to unify {len(unification_failed_curies)} CURIEs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = []\n",
    "q_node_id_keys = set([\"object\", \"subject\"])\n",
    "q_node_ids = []\n",
    "q_edge_ids = []\n",
    "for q_edge_id,edge_v in query[\"edges\"].items():\n",
    "    q_edge_ids.append(q_edge_id)\n",
    "    q_node_id_found = False\n",
    "    for k,v in edge_v.items():\n",
    "        if (k in q_node_id_keys) and (type(v) is str):\n",
    "            if v not in columns:\n",
    "                q_node_ids.append(v)\n",
    "                columns.append(v)\n",
    "            if not q_node_id_found:\n",
    "                q_node_id_found = True\n",
    "                columns.append(q_edge_id)\n",
    "\n",
    "query_nodes_with_ids = set()\n",
    "for k, v in query['nodes'].items():\n",
    "    node_ids = v.get('ids', [])\n",
    "    if node_ids:\n",
    "        query_nodes_with_ids.add(k)\n",
    "\n",
    "for q_node_id in q_node_ids:\n",
    "    columns.append(f\"{q_node_id}_original_curie\")\n",
    "    columns.append(f\"{q_node_id}_unified_curie\")\n",
    "\n",
    "columns.append(\"unified_curie_set\")\n",
    "\n",
    "trapi_result_columns = []\n",
    "for i in range(len(q_node_ids)):\n",
    "    trapi_result_columns.append(q_node_ids[i])\n",
    "    if i < len(q_edge_ids):\n",
    "        trapi_result_columns.append(q_edge_ids[i])\n",
    "\n",
    "unified_curie_columns = []\n",
    "for q_node_id in q_node_ids:\n",
    "    unified_curie_columns.append(f\"{q_node_id}_unified_curie\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "warning: predicate direction(s) may be switched\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n1</th>\n",
       "      <th>e0</th>\n",
       "      <th>n0</th>\n",
       "      <th>n2</th>\n",
       "      <th>e1</th>\n",
       "      <th>n1_original_curie</th>\n",
       "      <th>n1_unified_curie</th>\n",
       "      <th>n0_original_curie</th>\n",
       "      <th>n0_unified_curie</th>\n",
       "      <th>n2_original_curie</th>\n",
       "      <th>n2_unified_curie</th>\n",
       "      <th>unified_curie_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VEGFA</td>\n",
       "      <td>decreases_secretion_of</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>NCBIGene:7422</td>\n",
       "      <td>NCBIGene:7422</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{NCBIGene:7422, MESH:D001249, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VEGFA</td>\n",
       "      <td>decreases_secretion_of</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>contributes_to</td>\n",
       "      <td>NCBIGene:7422</td>\n",
       "      <td>NCBIGene:7422</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{NCBIGene:7422, MESH:D001249, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VEGFA</td>\n",
       "      <td>decreases_secretion_of</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>associated_with</td>\n",
       "      <td>NCBIGene:7422</td>\n",
       "      <td>NCBIGene:7422</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{NCBIGene:7422, MESH:D001249, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VEGFA</td>\n",
       "      <td>increases_secretion_of</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>NCBIGene:7422</td>\n",
       "      <td>NCBIGene:7422</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{NCBIGene:7422, MESH:D001249, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VEGFA</td>\n",
       "      <td>increases_secretion_of</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>contributes_to</td>\n",
       "      <td>NCBIGene:7422</td>\n",
       "      <td>NCBIGene:7422</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{NCBIGene:7422, MESH:D001249, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>VEGFA</td>\n",
       "      <td>increases_secretion_of</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>associated_with</td>\n",
       "      <td>NCBIGene:7422</td>\n",
       "      <td>NCBIGene:7422</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{NCBIGene:7422, MESH:D001249, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CASP8</td>\n",
       "      <td>increases_activity_of</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>NCBIGene:841</td>\n",
       "      <td>NCBIGene:841</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{NCBIGene:841, MESH:D001249, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CASP8</td>\n",
       "      <td>increases_activity_of</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>associated_with</td>\n",
       "      <td>NCBIGene:841</td>\n",
       "      <td>NCBIGene:841</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{NCBIGene:841, MESH:D001249, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>LYN</td>\n",
       "      <td>response_affected_by</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>associated_with</td>\n",
       "      <td>NCBIGene:4067</td>\n",
       "      <td>NCBIGene:4067</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{NCBIGene:4067, MESH:D001249, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>LYN</td>\n",
       "      <td>response_decreased_by</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>associated_with</td>\n",
       "      <td>NCBIGene:4067</td>\n",
       "      <td>NCBIGene:4067</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{NCBIGene:4067, MESH:D001249, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CXCL8</td>\n",
       "      <td>increases_secretion_of</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>NCBIGene:3576</td>\n",
       "      <td>NCBIGene:3576</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{MESH:D001249, NCBIGene:3576, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CXCL8</td>\n",
       "      <td>increases_secretion_of</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>contributes_to</td>\n",
       "      <td>NCBIGene:3576</td>\n",
       "      <td>NCBIGene:3576</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{MESH:D001249, NCBIGene:3576, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CXCL8</td>\n",
       "      <td>increases_secretion_of</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>occurs_together_in_literature_with</td>\n",
       "      <td>NCBIGene:3576</td>\n",
       "      <td>NCBIGene:3576</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{MESH:D001249, NCBIGene:3576, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PARP1</td>\n",
       "      <td>increases_degradation_of</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>occurs_together_in_literature_with</td>\n",
       "      <td>NCBIGene:142</td>\n",
       "      <td>NCBIGene:142</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{NCBIGene:142, MESH:D001249, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>response_increased_by</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>NCBIGene:5156</td>\n",
       "      <td>NCBIGene:5156</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{NCBIGene:5156, MESH:D001249, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>response_affected_by</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>NCBIGene:5156</td>\n",
       "      <td>NCBIGene:5156</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{NCBIGene:5156, MESH:D001249, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>decreases_expression_of</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>NCBIGene:5156</td>\n",
       "      <td>NCBIGene:5156</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{NCBIGene:5156, MESH:D001249, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>JUN</td>\n",
       "      <td>increases_activity_of</td>\n",
       "      <td>Imatinib mesylate</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>occurs_together_in_literature_with</td>\n",
       "      <td>NCBIGene:3725</td>\n",
       "      <td>NCBIGene:3725</td>\n",
       "      <td>PUBCHEM.COMPOUND:123596</td>\n",
       "      <td>MESH:D000068877</td>\n",
       "      <td>HP:0002099</td>\n",
       "      <td>MESH:D001249</td>\n",
       "      <td>{MESH:D001249, NCBIGene:3725, MESH:D000068877}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        n1                        e0                 n0      n2  \\\n",
       "0    VEGFA    decreases_secretion_of  Imatinib mesylate  Asthma   \n",
       "1    VEGFA    decreases_secretion_of  Imatinib mesylate  Asthma   \n",
       "2    VEGFA    decreases_secretion_of  Imatinib mesylate  Asthma   \n",
       "3    VEGFA    increases_secretion_of  Imatinib mesylate  Asthma   \n",
       "4    VEGFA    increases_secretion_of  Imatinib mesylate  Asthma   \n",
       "5    VEGFA    increases_secretion_of  Imatinib mesylate  Asthma   \n",
       "6    CASP8     increases_activity_of  Imatinib mesylate  Asthma   \n",
       "7    CASP8     increases_activity_of  Imatinib mesylate  Asthma   \n",
       "8      LYN      response_affected_by  Imatinib mesylate  Asthma   \n",
       "9      LYN     response_decreased_by  Imatinib mesylate  Asthma   \n",
       "10   CXCL8    increases_secretion_of  Imatinib mesylate  Asthma   \n",
       "11   CXCL8    increases_secretion_of  Imatinib mesylate  Asthma   \n",
       "12   CXCL8    increases_secretion_of  Imatinib mesylate  Asthma   \n",
       "13   PARP1  increases_degradation_of  Imatinib mesylate  Asthma   \n",
       "14  PDGFRA     response_increased_by  Imatinib mesylate  Asthma   \n",
       "15  PDGFRA      response_affected_by  Imatinib mesylate  Asthma   \n",
       "16  PDGFRA   decreases_expression_of  Imatinib mesylate  Asthma   \n",
       "17     JUN     increases_activity_of  Imatinib mesylate  Asthma   \n",
       "\n",
       "                                    e1 n1_original_curie n1_unified_curie  \\\n",
       "0       gene_associated_with_condition     NCBIGene:7422    NCBIGene:7422   \n",
       "1                       contributes_to     NCBIGene:7422    NCBIGene:7422   \n",
       "2                      associated_with     NCBIGene:7422    NCBIGene:7422   \n",
       "3       gene_associated_with_condition     NCBIGene:7422    NCBIGene:7422   \n",
       "4                       contributes_to     NCBIGene:7422    NCBIGene:7422   \n",
       "5                      associated_with     NCBIGene:7422    NCBIGene:7422   \n",
       "6       gene_associated_with_condition      NCBIGene:841     NCBIGene:841   \n",
       "7                      associated_with      NCBIGene:841     NCBIGene:841   \n",
       "8                      associated_with     NCBIGene:4067    NCBIGene:4067   \n",
       "9                      associated_with     NCBIGene:4067    NCBIGene:4067   \n",
       "10      gene_associated_with_condition     NCBIGene:3576    NCBIGene:3576   \n",
       "11                      contributes_to     NCBIGene:3576    NCBIGene:3576   \n",
       "12  occurs_together_in_literature_with     NCBIGene:3576    NCBIGene:3576   \n",
       "13  occurs_together_in_literature_with      NCBIGene:142     NCBIGene:142   \n",
       "14      gene_associated_with_condition     NCBIGene:5156    NCBIGene:5156   \n",
       "15      gene_associated_with_condition     NCBIGene:5156    NCBIGene:5156   \n",
       "16      gene_associated_with_condition     NCBIGene:5156    NCBIGene:5156   \n",
       "17  occurs_together_in_literature_with     NCBIGene:3725    NCBIGene:3725   \n",
       "\n",
       "          n0_original_curie n0_unified_curie n2_original_curie  \\\n",
       "0   PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "1   PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "2   PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "3   PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "4   PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "5   PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "6   PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "7   PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "8   PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "9   PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "10  PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "11  PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "12  PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "13  PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "14  PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "15  PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "16  PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "17  PUBCHEM.COMPOUND:123596  MESH:D000068877        HP:0002099   \n",
       "\n",
       "   n2_unified_curie                               unified_curie_set  \n",
       "0      MESH:D001249  {NCBIGene:7422, MESH:D001249, MESH:D000068877}  \n",
       "1      MESH:D001249  {NCBIGene:7422, MESH:D001249, MESH:D000068877}  \n",
       "2      MESH:D001249  {NCBIGene:7422, MESH:D001249, MESH:D000068877}  \n",
       "3      MESH:D001249  {NCBIGene:7422, MESH:D001249, MESH:D000068877}  \n",
       "4      MESH:D001249  {NCBIGene:7422, MESH:D001249, MESH:D000068877}  \n",
       "5      MESH:D001249  {NCBIGene:7422, MESH:D001249, MESH:D000068877}  \n",
       "6      MESH:D001249   {NCBIGene:841, MESH:D001249, MESH:D000068877}  \n",
       "7      MESH:D001249   {NCBIGene:841, MESH:D001249, MESH:D000068877}  \n",
       "8      MESH:D001249  {NCBIGene:4067, MESH:D001249, MESH:D000068877}  \n",
       "9      MESH:D001249  {NCBIGene:4067, MESH:D001249, MESH:D000068877}  \n",
       "10     MESH:D001249  {MESH:D001249, NCBIGene:3576, MESH:D000068877}  \n",
       "11     MESH:D001249  {MESH:D001249, NCBIGene:3576, MESH:D000068877}  \n",
       "12     MESH:D001249  {MESH:D001249, NCBIGene:3576, MESH:D000068877}  \n",
       "13     MESH:D001249   {NCBIGene:142, MESH:D001249, MESH:D000068877}  \n",
       "14     MESH:D001249  {NCBIGene:5156, MESH:D001249, MESH:D000068877}  \n",
       "15     MESH:D001249  {NCBIGene:5156, MESH:D001249, MESH:D000068877}  \n",
       "16     MESH:D001249  {NCBIGene:5156, MESH:D001249, MESH:D000068877}  \n",
       "17     MESH:D001249  {MESH:D001249, NCBIGene:3725, MESH:D000068877}  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trapi_results = trapi_message[\"results\"]\n",
    "\n",
    "#side exporation\n",
    "#additonal edge information from the knowledge graph might be explored for filtering later\n",
    "result_row_data = []\n",
    "for trapi_result in trapi_results:\n",
    "    curie_to_qnode_ids = dict()\n",
    "    for qnode_id, entries in trapi_result[\"node_bindings\"].items():\n",
    "        for entry in entries:\n",
    "            curie = entry[\"id\"]\n",
    "            if curie not in curie_to_qnode_ids:\n",
    "                curie_to_qnode_ids[curie] = []\n",
    "            curie_to_qnode_ids[curie].append(qnode_id)\n",
    "\n",
    "    row_data_template = dict()\n",
    "    q_edge_id_to_predicates = dict()\n",
    "    trapi_result_curie_set = set()\n",
    "    for qedge_id, entries in trapi_result[\"edge_bindings\"].items():\n",
    "        for entry in entries:\n",
    "            curie = entry[\"id\"]\n",
    "            kg_entry = trapi_message[\"knowledge_graph\"][\"edges\"][curie]\n",
    "            subject_curie = kg_entry[\"subject\"]\n",
    "            object_curie = kg_entry[\"object\"]\n",
    "            predicate_curie = kg_entry[\"predicate\"]\n",
    "            [predicate_prefix, predicate_identifier] = predicate_curie.split(\":\")\n",
    "\n",
    "            if qedge_id not in q_edge_id_to_predicates:\n",
    "                q_edge_id_to_predicates[qedge_id] = set()\n",
    "            q_edge_id_to_predicates[qedge_id].add(predicate_identifier)\n",
    "\n",
    "            for curie in [subject_curie, object_curie]:\n",
    "                for qnode_id in curie_to_qnode_ids[curie]:\n",
    "                    if curie in curie_to_unified_curie:\n",
    "                        unified_curie = curie_to_unified_curie[curie]\n",
    "                    else:\n",
    "                        break\n",
    "\n",
    "                    name = curie_to_name[curie]\n",
    "                    row_data_template[qnode_id] = name\n",
    "\n",
    "                    trapi_result_curie_set.add(unified_curie)\n",
    "                    row_data_template[qnode_id + \"_original_curie\"] = curie\n",
    "                    row_data_template[qnode_id + \"_unified_curie\"] = unified_curie\n",
    "\n",
    "    if len(trapi_result_curie_set) != len(q_node_ids):\n",
    "        #print(f'skipping {list(curie_to_qnode_ids.keys())}')\n",
    "        continue\n",
    "\n",
    "    row_data_template[\"unified_curie_set\"] = trapi_result_curie_set\n",
    "    q_edge_ids_processed = set()\n",
    "    row_datas = [row_data_template]\n",
    "    for q_edge_id,predicates in q_edge_id_to_predicates.items():\n",
    "        next_row_datas = []\n",
    "        for row_data in row_datas:\n",
    "            for predicate in predicates:\n",
    "                next_row_data = deepcopy(row_data)\n",
    "                next_row_data[q_edge_id] = predicate\n",
    "                next_row_datas.append(\n",
    "                    next_row_data \n",
    "                )\n",
    "        row_datas = next_row_datas\n",
    "    result_row_data += row_datas\n",
    "\n",
    "print(\"warning: predicate direction(s) may be switched\")\n",
    "trapi_results_df = pd.DataFrame.from_records(result_row_data, columns=columns)\n",
    "trapi_results_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identify result entities using user specified parameters\n",
    "\n",
    "```node_ID_for_grouping``` variable defines which node ID(s) from message.query_graph.nodes should be used for grouping. The entities mapped to that node ID are the \"result entities\". If \"all\" then use all the node ids in query are used. If specific node id set (e.g. {'MESH:D000068877', 'MESH:D001249'}), then use only these node ids are used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'MESH:D000068877', 'MESH:D001249'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#get all nodes with ids from the query \n",
    "user_specified_ids = set()\n",
    "for query_node in query_nodes_with_ids:\n",
    "    user_specified_ids.update(trapi_results_df[f'{query_node}_unified_curie'].tolist())\n",
    "user_specified_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'MESH:D000068877', 'MESH:D001249'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# identify node ids for identifying result entities\n",
    "if(node_ID_for_grouping != \"all\"):\n",
    "    user_specified_ids = user_specified_ids.intersection(node_ID_for_grouping)\n",
    "\n",
    "user_specified_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique CURIE count per query node:\n",
      "n1: 7\n",
      "n0: 1\n",
      "n2: 1\n"
     ]
    }
   ],
   "source": [
    "print(\"Unique CURIE count per query node:\")\n",
    "for q_node_id in q_node_ids:\n",
    "    print(f'{q_node_id}: {len(set(trapi_results_df[q_node_id]))}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get PFOCR Data\n",
    "\n",
    "Get PFOCR data from Dropbox."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figure_id</th>\n",
       "      <th>chemical_curie_set</th>\n",
       "      <th>disease_curie_set</th>\n",
       "      <th>gene_curie_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC5732092__cshperspect-CYT-028522_F2.jpg</td>\n",
       "      <td>set()</td>\n",
       "      <td>set()</td>\n",
       "      <td>{'NCBIGene:5595', 'NCBIGene:2919', 'NCBIGene:1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC5793760__cshperspect-TGF-022210_F4.jpg</td>\n",
       "      <td>set()</td>\n",
       "      <td>set()</td>\n",
       "      <td>{'NCBIGene:7042', 'NCBIGene:151449', 'NCBIGene...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC5793761__cshperspect-TGF-031989_F1.jpg</td>\n",
       "      <td>set()</td>\n",
       "      <td>set()</td>\n",
       "      <td>{'NCBIGene:4093', 'NCBIGene:5595', 'NCBIGene:1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC5830892__cshperspect-CEL-027961_F2.jpg</td>\n",
       "      <td>set()</td>\n",
       "      <td>set()</td>\n",
       "      <td>{'NCBIGene:72', 'NCBIGene:59', 'NCBIGene:55249...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC5830900__cshperspect-TGF-031997_F1.jpg</td>\n",
       "      <td>{'MESH:D011374'}</td>\n",
       "      <td>set()</td>\n",
       "      <td>{'NCBIGene:4093', 'NCBIGene:7042', 'NCBIGene:1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77714</th>\n",
       "      <td>PMC2804790__253_2009_2262_Fig1_HTML.jpg</td>\n",
       "      <td>{'MESH:D009243', 'MESH:C031105', 'MESH:D000447...</td>\n",
       "      <td>set()</td>\n",
       "      <td>{'NCBIGene:125', 'NCBIGene:137872', 'NCBIGene:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77715</th>\n",
       "      <td>PMC6332787__thnov09p0126g006.jpg</td>\n",
       "      <td>set()</td>\n",
       "      <td>set()</td>\n",
       "      <td>{'NCBIGene:6387', 'NCBIGene:8295', 'NCBIGene:2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77716</th>\n",
       "      <td>PMC5807036__IJO-52-03-0787-g01.jpg</td>\n",
       "      <td>set()</td>\n",
       "      <td>set()</td>\n",
       "      <td>{'NCBIGene:1977', 'NCBIGene:4254', 'NCBIGene:4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77717</th>\n",
       "      <td>PMC6770832__cancers-11-01236-g005.jpg</td>\n",
       "      <td>set()</td>\n",
       "      <td>set()</td>\n",
       "      <td>{'NCBIGene:5595', 'NCBIGene:4093', 'NCBIGene:7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77718</th>\n",
       "      <td>PMC4609065__12885_2015_1721_Fig5_HTML.jpg</td>\n",
       "      <td>{'MESH:D014443', 'MESH:C029269'}</td>\n",
       "      <td>set()</td>\n",
       "      <td>{'NCBIGene:2065', 'NCBIGene:2064', 'NCBIGene:2...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77719 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       figure_id  \\\n",
       "0      PMC5732092__cshperspect-CYT-028522_F2.jpg   \n",
       "1      PMC5793760__cshperspect-TGF-022210_F4.jpg   \n",
       "2      PMC5793761__cshperspect-TGF-031989_F1.jpg   \n",
       "3      PMC5830892__cshperspect-CEL-027961_F2.jpg   \n",
       "4      PMC5830900__cshperspect-TGF-031997_F1.jpg   \n",
       "...                                          ...   \n",
       "77714    PMC2804790__253_2009_2262_Fig1_HTML.jpg   \n",
       "77715           PMC6332787__thnov09p0126g006.jpg   \n",
       "77716         PMC5807036__IJO-52-03-0787-g01.jpg   \n",
       "77717      PMC6770832__cancers-11-01236-g005.jpg   \n",
       "77718  PMC4609065__12885_2015_1721_Fig5_HTML.jpg   \n",
       "\n",
       "                                      chemical_curie_set disease_curie_set  \\\n",
       "0                                                  set()             set()   \n",
       "1                                                  set()             set()   \n",
       "2                                                  set()             set()   \n",
       "3                                                  set()             set()   \n",
       "4                                       {'MESH:D011374'}             set()   \n",
       "...                                                  ...               ...   \n",
       "77714  {'MESH:D009243', 'MESH:C031105', 'MESH:D000447...             set()   \n",
       "77715                                              set()             set()   \n",
       "77716                                              set()             set()   \n",
       "77717                                              set()             set()   \n",
       "77718                   {'MESH:D014443', 'MESH:C029269'}             set()   \n",
       "\n",
       "                                          gene_curie_set  \n",
       "0      {'NCBIGene:5595', 'NCBIGene:2919', 'NCBIGene:1...  \n",
       "1      {'NCBIGene:7042', 'NCBIGene:151449', 'NCBIGene...  \n",
       "2      {'NCBIGene:4093', 'NCBIGene:5595', 'NCBIGene:1...  \n",
       "3      {'NCBIGene:72', 'NCBIGene:59', 'NCBIGene:55249...  \n",
       "4      {'NCBIGene:4093', 'NCBIGene:7042', 'NCBIGene:1...  \n",
       "...                                                  ...  \n",
       "77714  {'NCBIGene:125', 'NCBIGene:137872', 'NCBIGene:...  \n",
       "77715  {'NCBIGene:6387', 'NCBIGene:8295', 'NCBIGene:2...  \n",
       "77716  {'NCBIGene:1977', 'NCBIGene:4254', 'NCBIGene:4...  \n",
       "77717  {'NCBIGene:5595', 'NCBIGene:4093', 'NCBIGene:7...  \n",
       "77718  {'NCBIGene:2065', 'NCBIGene:2064', 'NCBIGene:2...  \n",
       "\n",
       "[77719 rows x 4 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#create a data frame for all the figure data from PFOCR\n",
    "#expected size of final data frame (as in Jan, 2023): 77719 rows × 4 columns\n",
    "figures_df = pd.read_csv('https://www.dropbox.com/s/3eukom49hqskp88/figures_dataframe.csv?dl=1')\n",
    "figures_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figure_url</th>\n",
       "      <th>figure_title</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>figure_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PMC5732092__cshperspect-CYT-028522_F2.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "      <td>Interleukin (IL)-17RA/RC signaling pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC5793760__cshperspect-TGF-022210_F4.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "      <td>TGFB signaling pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC5793761__cshperspect-TGF-031989_F1.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "      <td>Bone morphogenetic protein (BMP) signaling pat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC5830892__cshperspect-CEL-027961_F2.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "      <td>FERM-binding partners of Crumbs3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC5830900__cshperspect-TGF-031997_F1.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "      <td>Role of the TGFB family in mammary gland devel...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC2804790__253_2009_2262_Fig1_HTML.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...</td>\n",
       "      <td>Ehrlich degradation pathway from amino acid to...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC6332787__thnov09p0126g006.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "      <td>Illustration of the mechanisms underlying LPD ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC5807036__IJO-52-03-0787-g01.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "      <td>Tumorigenic proteins significantly upregulated...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC6770832__cancers-11-01236-g005.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "      <td>Simplified TGFB pathway leading to p21 expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC4609065__12885_2015_1721_Fig5_HTML.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4...</td>\n",
       "      <td>EGFR and ANXA3 to fluoropyrimidine resistance ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77719 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                  figure_url  \\\n",
       "figure_id                                                                                      \n",
       "PMC5732092__cshperspect-CYT-028522_F2.jpg  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...   \n",
       "PMC5793760__cshperspect-TGF-022210_F4.jpg  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...   \n",
       "PMC5793761__cshperspect-TGF-031989_F1.jpg  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...   \n",
       "PMC5830892__cshperspect-CEL-027961_F2.jpg  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...   \n",
       "PMC5830900__cshperspect-TGF-031997_F1.jpg  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...   \n",
       "...                                                                                      ...   \n",
       "PMC2804790__253_2009_2262_Fig1_HTML.jpg    https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...   \n",
       "PMC6332787__thnov09p0126g006.jpg           https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...   \n",
       "PMC5807036__IJO-52-03-0787-g01.jpg         https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...   \n",
       "PMC6770832__cancers-11-01236-g005.jpg      https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...   \n",
       "PMC4609065__12885_2015_1721_Fig5_HTML.jpg  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4...   \n",
       "\n",
       "                                                                                figure_title  \n",
       "figure_id                                                                                     \n",
       "PMC5732092__cshperspect-CYT-028522_F2.jpg        Interleukin (IL)-17RA/RC signaling pathways  \n",
       "PMC5793760__cshperspect-TGF-022210_F4.jpg                            TGFB signaling pathways  \n",
       "PMC5793761__cshperspect-TGF-031989_F1.jpg  Bone morphogenetic protein (BMP) signaling pat...  \n",
       "PMC5830892__cshperspect-CEL-027961_F2.jpg                   FERM-binding partners of Crumbs3  \n",
       "PMC5830900__cshperspect-TGF-031997_F1.jpg  Role of the TGFB family in mammary gland devel...  \n",
       "...                                                                                      ...  \n",
       "PMC2804790__253_2009_2262_Fig1_HTML.jpg    Ehrlich degradation pathway from amino acid to...  \n",
       "PMC6332787__thnov09p0126g006.jpg           Illustration of the mechanisms underlying LPD ...  \n",
       "PMC5807036__IJO-52-03-0787-g01.jpg         Tumorigenic proteins significantly upregulated...  \n",
       "PMC6770832__cancers-11-01236-g005.jpg      Simplified TGFB pathway leading to p21 expression  \n",
       "PMC4609065__12885_2015_1721_Fig5_HTML.jpg  EGFR and ANXA3 to fluoropyrimidine resistance ...  \n",
       "\n",
       "[77719 rows x 2 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in figure metadata for all PFOCR figures\n",
    "figure_metadata_df = pd.read_csv('https://www.dropbox.com/s/9jorbzpq2k8n5tr/figures_metadata.csv?dl=1', index_col=0)\n",
    "figure_metadata_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figure_id</th>\n",
       "      <th>curie</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC5732092__cshperspect-CYT-028522_F2.jpg</td>\n",
       "      <td>NCBIGene:5601</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC5732092__cshperspect-CYT-028522_F2.jpg</td>\n",
       "      <td>NCBIGene:5595</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC5732092__cshperspect-CYT-028522_F2.jpg</td>\n",
       "      <td>NCBIGene:1051</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC5732092__cshperspect-CYT-028522_F2.jpg</td>\n",
       "      <td>NCBIGene:10131</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC5732092__cshperspect-CYT-028522_F2.jpg</td>\n",
       "      <td>NCBIGene:23765</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1665596</th>\n",
       "      <td>PMC4609065__12885_2015_1721_Fig5_HTML.jpg</td>\n",
       "      <td>NCBIGene:1956</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1665597</th>\n",
       "      <td>PMC4609065__12885_2015_1721_Fig5_HTML.jpg</td>\n",
       "      <td>NCBIGene:4893</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1665598</th>\n",
       "      <td>PMC4609065__12885_2015_1721_Fig5_HTML.jpg</td>\n",
       "      <td>NCBIGene:2064</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1665599</th>\n",
       "      <td>PMC4609065__12885_2015_1721_Fig5_HTML.jpg</td>\n",
       "      <td>NCBIGene:3845</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1665600</th>\n",
       "      <td>PMC4609065__12885_2015_1721_Fig5_HTML.jpg</td>\n",
       "      <td>NCBIGene:306</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1665601 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         figure_id           curie category\n",
       "0        PMC5732092__cshperspect-CYT-028522_F2.jpg   NCBIGene:5601     gene\n",
       "1        PMC5732092__cshperspect-CYT-028522_F2.jpg   NCBIGene:5595     gene\n",
       "2        PMC5732092__cshperspect-CYT-028522_F2.jpg   NCBIGene:1051     gene\n",
       "3        PMC5732092__cshperspect-CYT-028522_F2.jpg  NCBIGene:10131     gene\n",
       "4        PMC5732092__cshperspect-CYT-028522_F2.jpg  NCBIGene:23765     gene\n",
       "...                                            ...             ...      ...\n",
       "1665596  PMC4609065__12885_2015_1721_Fig5_HTML.jpg   NCBIGene:1956     gene\n",
       "1665597  PMC4609065__12885_2015_1721_Fig5_HTML.jpg   NCBIGene:4893     gene\n",
       "1665598  PMC4609065__12885_2015_1721_Fig5_HTML.jpg   NCBIGene:2064     gene\n",
       "1665599  PMC4609065__12885_2015_1721_Fig5_HTML.jpg   NCBIGene:3845     gene\n",
       "1665600  PMC4609065__12885_2015_1721_Fig5_HTML.jpg    NCBIGene:306     gene\n",
       "\n",
       "[1665601 rows x 3 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in figure CURIES for all PFOCR figures\n",
    "figure_curie_df = pd.read_csv('https://www.dropbox.com/s/y3ks74obs3nrjjz/figures_curie.csv?dl=1')\n",
    "figure_curie_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Enrichment/grouping of TRAPI results using PFOCR figures\n",
    "\n",
    "From discussion with Andrew, we want to first identify a set of figures that can serve to cluster the TRAPI results. We can call these the \"cluster figures\".\n",
    "\n",
    "**Identify Cluster Figures**\n",
    "\n",
    "We identify the top cluster figures by iteratively running Fisher's Exact Test to identify PFOCR figures that are most similar to all the CURIEs from the BTE TRAPI results. To get each cluster figure, we do the following:\n",
    "\n",
    "1. Run Fisher's Exact Test for the CURIEs from each figure vs. the set of the unique CURIEs found from all the TRAPI results, limiting each  set of CURIEs to those with categories found in both PFOCR and BTE TRAPI, e.g., \"disease\" is found in PFOCR and \"biolink:Disease\" is found in BTE TRAPI\n",
    "2. Select the figure with the lowest p-value as a cluster figure\n",
    "3. Exclude that figure and the CURIEs in it\n",
    "4. Repeat to get the next cluster figure, continuing until:\n",
    "    - we run out of BTE TRAPI results\n",
    "    - we run out of PFOCR figures\n",
    "    - we've repeated steps 1-3 a user-defined number of times (varibale ```n```)\n",
    "\n",
    "**Cluster BTE TRAPI Results by Selected PFOCR Figures (Cluster Figures)**\n",
    "\n",
    "Once we have a set of cluster figures, we use them to cluster the BTE TRAPI results. We do this by running Fisher's Exact Test to identify which Cluster Figure each BTE TRAPI result is most similar to. We don't actually cluster the TRAPI results by figure, however, because if a result is highly relevant to two figures, it can show up in both. It's more accurate to think of the figure as being a different view of the TRAPI results.\n",
    "\n",
    "**View TRAPI Results by PFOCR Figure**\n",
    "\n",
    "With BTE TRAPI results \"clustered\" (not actually clustered, because we show the same result multiple times for different figures if it's closely related to both) by PFOCR figure, we display the top figures, each with their most closely related BTE TRAPI results.\n",
    "\n",
    "**Latest from Andrew**\n",
    "Implementation details:\n",
    "The enrichment analysis will be iterative in design. On the first iteration, all result entities will be grouped in one list, and an enrichment analysis will be run relative to all PFOCR figures. The most significantly enriched PFOCR figure will be set aside as PFOCR Figure 1.\n",
    "The result entities that occurred in PFOCR Figure 1 will be removed from the result entity list, and a new enrichment analysis will be performed. The most significantly-enriched PFOCR figure will be set aside as PFOCR Figure 2.\n",
    "The process will be repeated a total of n times.\n",
    "The basic flow has been worked out in https://github.com/wikipathways/pathway-figure-ocr/blob/master/notebooks/bte_clustering.ipynb, but that notebook needs to be simplified for broader use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find the corresponding PFOCR category for a biolink category\n",
    "biolink_category_to_pfocr_category = {\n",
    "    \"biolink:Drug\": \"chemical\",\n",
    "    \"biolink:ChemicalEntity\": \"chemical\",\n",
    "    \"biolink:SmallMolecule\": \"chemical\",\n",
    "    \"biolink:Disease\": \"disease\",\n",
    "    \"biolink:DiseaseOrPhenotypicFeature\": \"disease\",\n",
    "    \"biolink:Gene\": \"gene\",\n",
    "    \"biolink:GeneProduct\": \"gene\",\n",
    "    \"biolink:Protein\": \"gene\",\n",
    "}\n",
    "\n",
    "# limit to categories returned from query\n",
    "relevant_biolink_categories = set(biolink_category_to_pfocr_category.keys()).intersection(curie_categories)\n",
    "relevant_pfocr_categories = set([\n",
    "    biolink_category_to_pfocr_category[\n",
    "        biolink_category\n",
    "    ] for biolink_category in relevant_biolink_categories\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importing reduce for \n",
    "# rolling computations\n",
    "from functools import reduce\n",
    "  \n",
    "def set_union(series):\n",
    "    return reduce(lambda x, y: x | y, series)\n",
    "\n",
    "def create_set(series):\n",
    "    return set(series.tolist())\n",
    "\n",
    "\n",
    "# TODO: remove cluster_figure_data and its associated DF.\n",
    "# It's just a temporary thing to show what's happening inside.\n",
    "cluster_figure_data = []\n",
    "def get_next_cluster_figures(\n",
    "    category_matched_pfocr_curie_count,\n",
    "    remaining_figures_df,\n",
    "    remaining_trapi_curie_count,\n",
    "    required_curies,\n",
    "    iteration_limit=n, #this is the user-defined parameter for the number of figures\n",
    "    i=0,\n",
    "    cluster_figure_ids=[],\n",
    "):\n",
    "    remaining_figure_count = len(\n",
    "        remaining_figures_df\n",
    "    )\n",
    "\n",
    "    if (\n",
    "        i >= iteration_limit\n",
    "    ) or (\n",
    "        len(remaining_trapi_curie_count) == 0\n",
    "    ) or (\n",
    "        remaining_figure_count == 0\n",
    "    ):\n",
    "        return cluster_figure_ids\n",
    "    \n",
    "    remaining_figures_df[\"figure_curie_count\"] = remaining_figures_df[\"curie_set\"].map(len)\n",
    "\n",
    "    remaining_figures_df[\"figure_curies_in_trapi_results\"] = remaining_figures_df[\"curie_set\"].map(\n",
    "        lambda curie_set: remaining_trapi_curie_count.intersection(\n",
    "            set(curie_set)\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # number of CURIEs in both TRAPI results and figure\n",
    "    remaining_figures_df[\"yes_trapi_results_yes_figure\"] = remaining_figures_df[\n",
    "        \"figure_curies_in_trapi_results\"\n",
    "    ].map(len)\n",
    "\n",
    "    # number of CURIEs in TRAPI results but not in figure\n",
    "    remaining_figures_df[\"yes_trapi_results_no_figure\"] = (\n",
    "        len(remaining_trapi_curie_count) -\n",
    "        remaining_figures_df[\"yes_trapi_results_yes_figure\"]\n",
    "    )\n",
    "\n",
    "    # number of CURIEs not in TRAPI results but in figure\n",
    "    remaining_figures_df[\"no_trapi_results_yes_figure\"] = (\n",
    "        remaining_figures_df[\"figure_curie_count\"] -\n",
    "        remaining_figures_df[\"yes_trapi_results_yes_figure\"]\n",
    "    )\n",
    "\n",
    "    # number of CURIEs not in TRAPI results and not in figure,\n",
    "    # ie., number of unique CURIEs only in other figures\n",
    "    remaining_figures_df[\"no_trapi_results_no_figure\"] = (\n",
    "        category_matched_pfocr_curie_count -\n",
    "        len(remaining_trapi_curie_count) -\n",
    "        remaining_figures_df[\"no_trapi_results_yes_figure\"]\n",
    "    )\n",
    "\n",
    "    # see https://stackoverflow.com/a/58661068/5354298\n",
    "    _, _, twosided = pvalue_npy(\n",
    "        remaining_figures_df[\"yes_trapi_results_yes_figure\"].to_numpy('uint'),\n",
    "        remaining_figures_df[\"yes_trapi_results_no_figure\"].to_numpy('uint'),\n",
    "        remaining_figures_df[\"no_trapi_results_yes_figure\"].to_numpy('uint'),\n",
    "        remaining_figures_df[\"no_trapi_results_no_figure\"].to_numpy('uint'),\n",
    "    )\n",
    "\n",
    "    remaining_figures_df[\"p_value\"] = pd.Series(twosided, index=remaining_figures_df.index)\n",
    "    remaining_figures_df[\"p_value\"].sort_values()\n",
    "    \n",
    "    min_df = remaining_figures_df[\n",
    "        remaining_figures_df[\"p_value\"] == remaining_figures_df[\"p_value\"].min()\n",
    "    ]\n",
    "    \n",
    "    if required_curies:\n",
    "        # TODO: it might make sense to keep prioritizing the required CURIEs, but\n",
    "        #       it's not clear whether this is always what we want.\n",
    "        #       In some cases, we could obscure a different set of results that may\n",
    "        #       be valuable but not because of the overlap with the required CURIEs.\n",
    "        curies_to_exclude = set(\n",
    "            min_df[\"figure_curies_in_trapi_results\"].agg(set_union)\n",
    "        ) - required_curies\n",
    "    else:\n",
    "        curies_to_exclude = set(min_df[\"figure_curies_in_trapi_results\"].agg(set_union))\n",
    "        \n",
    "    if len(curies_to_exclude - required_curies) == 0:\n",
    "        # On this iteration, we failed to match any new CURIE(s), other than\n",
    "        # possibly just the require CURIEs.\n",
    "        return cluster_figure_ids\n",
    "        \n",
    "    new_cluster_figure_ids = min_df.index.tolist()\n",
    "    cluster_figure_ids.extend(new_cluster_figure_ids)\n",
    "    \n",
    "    next_min_df_score = remaining_figures_df[\n",
    "        remaining_figures_df[\"p_value\"] > remaining_figures_df[\"p_value\"].min()\n",
    "    ][\"p_value\"].min()\n",
    "    next_min_df = remaining_figures_df[remaining_figures_df[\"p_value\"] == next_min_df_score]\n",
    "    \n",
    "    for cluster_figure_id, min_row in min_df.iterrows():\n",
    "        cluster_figure_data.append({\n",
    "            \"iteration\": i,\n",
    "            \"figure_id\": cluster_figure_id,\n",
    "            \"p_value\": min_row[\"p_value\"],\n",
    "            \"next_best_figure_ids\": next_min_df.index.tolist(),\n",
    "            \"next_best_p_value\": next_min_df[\"p_value\"].min(),\n",
    "            \"yes_trapi_results_yes_figure\": min_row[\"yes_trapi_results_yes_figure\"],\n",
    "            \"yes_trapi_results_no_figure\": min_row[\"yes_trapi_results_no_figure\"],\n",
    "            \"no_trapi_results_yes_figure\": min_row[\"no_trapi_results_yes_figure\"],\n",
    "            \"no_trapi_results_no_figure\": min_row[\"no_trapi_results_no_figure\"],\n",
    "            \"figure_curies_in_trapi_results\": min_row[\"figure_curies_in_trapi_results\"],\n",
    "            \"curie_set\": min_row[\"curie_set\"],\n",
    "            \"curies_to_exclude\": curies_to_exclude,\n",
    "            \"len_curies_to_exclude\": len(curies_to_exclude),\n",
    "            \"remaining_figure_count\": remaining_figure_count,\n",
    "            \"remaining_trapi_curie_count\": len(remaining_trapi_curie_count),\n",
    "            \"cluster_figure_id_count\": len(cluster_figure_ids),\n",
    "        })\n",
    "    \n",
    "    # for next iteration, exclude figures we've already added to cluster figures\n",
    "    next_remaining_figures_df = remaining_figures_df[\n",
    "        (~remaining_figures_df.index.isin(\n",
    "            cluster_figure_ids\n",
    "        ))\n",
    "    ]\n",
    "    \n",
    "    print(\n",
    "        f'iteration {i:>2}: figures: {len(new_cluster_figure_ids):>2}, CURIEs: {len(curies_to_exclude):>2}'\n",
    "    )\n",
    "    \n",
    "    return get_next_cluster_figures(\n",
    "        category_matched_pfocr_curie_count,\n",
    "        next_remaining_figures_df.reset_index().set_index(\"figure_id\"),\n",
    "        remaining_trapi_curie_count - curies_to_exclude,\n",
    "        required_curies,\n",
    "        iteration_limit,\n",
    "        i + 1,\n",
    "        cluster_figure_ids,\n",
    "    )\n",
    "\n",
    "category_matched_figure_curie_df = figure_curie_df[figure_curie_df[\"category\"].isin(\n",
    "    list(relevant_pfocr_categories)\n",
    ")]\n",
    "category_matched_pfocr_curie_count = len(set(\n",
    "    category_matched_figure_curie_df[\"curie\"].drop_duplicates()\n",
    "))\n",
    "print(f'category_matched_pfocr_curie_count: {category_matched_pfocr_curie_count}')\n",
    "\n",
    "category_matched_figures_df = category_matched_figure_curie_df[\n",
    "    [\"figure_id\", \"curie\"]\n",
    "].groupby(\"figure_id\").agg(\n",
    "    create_set\n",
    ").rename(columns={\n",
    "    \"curie\": \"curie_set\"\n",
    "})\n",
    "\n",
    "category_matched_figures_df[\"figure_curies_in_trapi_results_count\"] = category_matched_figures_df[\n",
    "    \"curie_set\"\n",
    "].map(\n",
    "    lambda curie_set: len(trapi_results_unified_curies.intersection(\n",
    "        set(curie_set)\n",
    "    ))\n",
    ")\n",
    "\n",
    "overlapping_category_matched_figures_df = category_matched_figures_df[\n",
    "    category_matched_figures_df[\"figure_curies_in_trapi_results_count\"] > 0\n",
    "]\n",
    "\n",
    "# Require that cluster figures have specific user-specified CURIEs\n",
    "if(required_curies == \"all\"):\n",
    "    required_curies = user_specified_ids\n",
    "if required_curies:\n",
    "    overlapping_category_matched_figures_df = overlapping_category_matched_figures_df[\n",
    "        overlapping_category_matched_figures_df[\"curie_set\"].map(\n",
    "            # require all of them\n",
    "            # lambda curie_set: len(curie_set.intersection(required_curies)) == len(required_curies)\n",
    "            # require at least one of them\n",
    "            lambda curie_set: len(curie_set.intersection(required_curies)) > 0\n",
    "        )\n",
    "    ]\n",
    "\n",
    "cluster_figure_ids_out = get_next_cluster_figures(\n",
    "    category_matched_pfocr_curie_count,\n",
    "    overlapping_category_matched_figures_df.reset_index().set_index(\"figure_id\"),\n",
    "    trapi_results_unified_curies,\n",
    "    required_curies,\n",
    ")\n",
    "print(f'cluster figure count: {len(cluster_figure_ids_out)}')\n",
    "cluster_figure_df = pd.DataFrame.from_records(cluster_figure_data)\n",
    "cluster_figure_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## View TRAPI Results by PFOCR Figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figures_df = category_matched_figure_curie_df[\n",
    "    category_matched_figure_curie_df[\"figure_id\"].isin(set(cluster_figure_df[\"figure_id\"]))\n",
    "][[\"figure_id\", \"curie\"]].groupby(\"figure_id\").agg(\n",
    "    create_set\n",
    ").rename(columns={\n",
    "    \"curie\": \"figure_curie_set\"\n",
    "}).join(\n",
    "    figure_metadata_df\n",
    ").reset_index()\n",
    "\n",
    "figures_df[\"figure_curie_count\"] = figures_df[\"figure_curie_set\"].map(len)\n",
    "\n",
    "figures_df['crossjoin_key'] = 0\n",
    "trapi_results_df['crossjoin_key'] = 0\n",
    "\n",
    "trapi_results_with_figures_df = trapi_results_df.merge(\n",
    "    figures_df, on='crossjoin_key', how='outer'\n",
    ").rename(columns={\n",
    "    \"unified_curie_set\": \"trapi_result_curie_set\"\n",
    "})\n",
    "\n",
    "trapi_results_with_figures_df[\"common_curie_set\"] = trapi_results_with_figures_df.apply(\n",
    "    lambda r: r[\"trapi_result_curie_set\"] & r[\"figure_curie_set\"], axis=1\n",
    ")\n",
    "trapi_results_with_figures_df\n",
    "# number of CURIEs in both TRAPI result and figure\n",
    "# trapi_curies_in_figure\n",
    "# x\n",
    "# Selected & Having the property\n",
    "trapi_results_with_figures_df[\"yes_trapi_result_yes_figure\"] = (\n",
    "    trapi_results_with_figures_df[\"common_curie_set\"].map(len)\n",
    ")\n",
    "\n",
    "# number of CURIEs in TRAPI result but not in figure\n",
    "# q_node_id_count - trapi_curies_in_figure\n",
    "# n - x\n",
    "# Selected & Not Having the property\n",
    "trapi_results_with_figures_df[\"yes_trapi_result_no_figure\"] = (\n",
    "    len(q_node_ids) - trapi_results_with_figures_df[\"yes_trapi_result_yes_figure\"]\n",
    ")\n",
    "\n",
    "# number of CURIEs not in TRAPI result but in figure\n",
    "# curies_in_figure - trapi_curies_in_figure\n",
    "# N - x\n",
    "# Not Selected & Having the property\n",
    "trapi_results_with_figures_df[\"no_trapi_result_yes_figure\"] = (\n",
    "    trapi_results_with_figures_df[\"figure_curie_set\"].map(len) - \n",
    "    trapi_results_with_figures_df[\"yes_trapi_result_yes_figure\"]\n",
    ")\n",
    "\n",
    "# number of CURIEs not in TRAPI result and not in figure,\n",
    "# ie., number of unique CURIEs only in other figures\n",
    "# 28735 - curies_in_figure - q_node_id_count + trapi_curies_in_figure\n",
    "# M - (n + N) + x\n",
    "# Not Selected & Not Having the property\n",
    "trapi_results_with_figures_df[\"no_trapi_result_no_figure\"] = (\n",
    "    category_matched_pfocr_curie_count -\n",
    "    trapi_results_with_figures_df[\"figure_curie_set\"].map(len) -\n",
    "    len(q_node_ids) +\n",
    "    trapi_results_with_figures_df[\"yes_trapi_result_yes_figure\"]\n",
    ")\n",
    "\n",
    "# see https://stackoverflow.com/a/58661068/5354298\n",
    "_, _, twosided = pvalue_npy(\n",
    "    trapi_results_with_figures_df[\"yes_trapi_result_yes_figure\"].to_numpy('uint'),\n",
    "    trapi_results_with_figures_df[\"yes_trapi_result_no_figure\"].to_numpy('uint'),\n",
    "    trapi_results_with_figures_df[\"no_trapi_result_yes_figure\"].to_numpy('uint'),\n",
    "    trapi_results_with_figures_df[\"no_trapi_result_no_figure\"].to_numpy('uint'),\n",
    ")\n",
    "\n",
    "trapi_results_with_figures_df[\"p_value\"] = pd.Series(twosided, index=trapi_results_with_figures_df.index)\n",
    "trapi_results_with_figures_df[\"p_value\"].sort_values()\n",
    "\n",
    "# Needed in order to use the TRAPI result CURIEs as a key. A Python Set is not hashable.\n",
    "trapi_results_with_figures_df[\"trapi_result_curies_key\"] = trapi_results_with_figures_df[\n",
    "    \"trapi_result_curie_set\"\n",
    "].map(\n",
    "    lambda x: tuple(sorted(x))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trapi_results_with_figures_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This isn't really a significance threshold. It's more just intended to\n",
    "# exclude results where the p-value is 1 or very close to it.\n",
    "threshold = 0.9\n",
    "significant_trapi_results_with_figures_df = trapi_results_with_figures_df[\n",
    "    trapi_results_with_figures_df[\"p_value\"] < 0.9\n",
    "]\n",
    "\n",
    "# all\n",
    "cooccurrence_df = significant_trapi_results_with_figures_df.sort_values(\n",
    "    \"p_value\"\n",
    ")\n",
    "\n",
    "### first and any other query node\n",
    "##q_node_id_combination = (q_node_ids[0],)\n",
    "\n",
    "## first and second query nodes\n",
    "##q_node_id_combination = (q_node_ids[0], q_node_ids[1])\n",
    "\n",
    "### first and last query nodes\n",
    "##q_node_id_combination = (q_node_ids[0], q_node_ids[-1])\n",
    "\n",
    "### second and last query nodes\n",
    "##q_node_id_combination = (q_node_ids[1], q_node_ids[-1])\n",
    "\n",
    "### every query node with 'ids' specified\n",
    "##q_node_id_combination = query_nodes_with_ids\n",
    "\n",
    "#print(f'{\" & \".join(q_node_id_combination)}')\n",
    "#s = pd.Series(\n",
    "#    [True] * len(significant_trapi_results_with_figures_df),\n",
    "#    index=significant_trapi_results_with_figures_dfindex\n",
    "#)\n",
    "#for q_node_id in q_node_id_combination:\n",
    "#    s = s & significant_trapi_results_with_figures_df.apply(\n",
    "#        lambda r: r[f'{q_node_id}_unified_curie'] in r[\"common_curie_set\"], axis=1\n",
    "#    )\n",
    "#cooccurrence_df = significant_trapi_results_with_figures_df[s == True].sort_values(\n",
    "#    \"p_value\"\n",
    "#)\n",
    "\n",
    "cooccurrence_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "authorship_tag": "ABX9TyP6xHxRuuT/oO2JrYOIGJq6",
   "include_colab_link": true,
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
